// @jest/fake-timers/build/legacyFakeTimers.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.default = void 0;
function _util() {
  const data = require('util');
  _util = function () {
    return data;
  };
  return data;
}
function _jestMessageUtil() {
  const data = require('jest-message-util');
  _jestMessageUtil = function () {
    return data;
  };
  return data;
}
function _jestUtil() {
  const data = require('jest-util');
  _jestUtil = function () {
    return data;
  };
  return data;
}
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable local/prefer-spread-eventually */

const MS_IN_A_YEAR = 31536000000;
class FakeTimers {
  _cancelledTicks;
  _config;
  _disposed;
  _fakeTimerAPIs;
  _fakingTime = false;
  _global;
  _immediates;
  _maxLoops;
  _moduleMocker;
  _now;
  _ticks;
  _timerAPIs;
  _timers;
  _uuidCounter;
  _timerConfig;
  constructor({global, moduleMocker, timerConfig, config, maxLoops}) {
    this._global = global;
    this._timerConfig = timerConfig;
    this._config = config;
    this._maxLoops = maxLoops || 100000;
    this._uuidCounter = 1;
    this._moduleMocker = moduleMocker;

    // Store original timer APIs for future reference
    this._timerAPIs = {
      cancelAnimationFrame: global.cancelAnimationFrame,
      clearImmediate: global.clearImmediate,
      clearInterval: global.clearInterval,
      clearTimeout: global.clearTimeout,
      nextTick: global.process && global.process.nextTick,
      requestAnimationFrame: global.requestAnimationFrame,
      setImmediate: global.setImmediate,
      setInterval: global.setInterval,
      setTimeout: global.setTimeout
    };
    this._disposed = false;
    this.reset();
  }
  clearAllTimers() {
    this._immediates = [];
    this._timers.clear();
  }
  dispose() {
    this._disposed = true;
    this.clearAllTimers();
  }
  reset() {
    this._cancelledTicks = {};
    this._now = 0;
    this._ticks = [];
    this._immediates = [];
    this._timers = new Map();
  }
  now() {
    if (this._fakingTime) {
      return this._now;
    }
    return Date.now();
  }
  runAllTicks() {
    this._checkFakeTimers();
    // Only run a generous number of ticks and then bail.
    // This is just to help avoid recursive loops
    let i;
    for (i = 0; i < this._maxLoops; i++) {
      const tick = this._ticks.shift();
      if (tick === undefined) {
        break;
      }
      if (
        !Object.prototype.hasOwnProperty.call(this._cancelledTicks, tick.uuid)
      ) {
        // Callback may throw, so update the map prior calling.
        this._cancelledTicks[tick.uuid] = true;
        tick.callback();
      }
    }
    if (i === this._maxLoops) {
      throw new Error(
        `Ran ${this._maxLoops} ticks, and there are still more! ` +
          "Assuming we've hit an infinite recursion and bailing out..."
      );
    }
  }
  runAllImmediates() {
    this._checkFakeTimers();
    // Only run a generous number of immediates and then bail.
    let i;
    for (i = 0; i < this._maxLoops; i++) {
      const immediate = this._immediates.shift();
      if (immediate === undefined) {
        break;
      }
      this._runImmediate(immediate);
    }
    if (i === this._maxLoops) {
      throw new Error(
        `Ran ${this._maxLoops} immediates, and there are still more! Assuming ` +
          "we've hit an infinite recursion and bailing out..."
      );
    }
  }
  _runImmediate(immediate) {
    try {
      immediate.callback();
    } finally {
      this._fakeClearImmediate(immediate.uuid);
    }
  }
  runAllTimers() {
    this._checkFakeTimers();
    this.runAllTicks();
    this.runAllImmediates();

    // Only run a generous number of timers and then bail.
    // This is just to help avoid recursive loops
    let i;
    for (i = 0; i < this._maxLoops; i++) {
      const nextTimerHandleAndExpiry = this._getNextTimerHandleAndExpiry();

      // If there are no more timer handles, stop!
      if (nextTimerHandleAndExpiry === null) {
        break;
      }
      const [nextTimerHandle, expiry] = nextTimerHandleAndExpiry;
      this._now = expiry;
      this._runTimerHandle(nextTimerHandle);

      // Some of the immediate calls could be enqueued
      // during the previous handling of the timers, we should
      // run them as well.
      if (this._immediates.length) {
        this.runAllImmediates();
      }
      if (this._ticks.length) {
        this.runAllTicks();
      }
    }
    if (i === this._maxLoops) {
      throw new Error(
        `Ran ${this._maxLoops} timers, and there are still more! ` +
          "Assuming we've hit an infinite recursion and bailing out..."
      );
    }
  }
  runOnlyPendingTimers() {
    // We need to hold the current shape of `this._timers` because existing
    // timers can add new ones to the map and hence would run more than necessary.
    // See https://github.com/jestjs/jest/pull/4608 for details
    const timerEntries = Array.from(this._timers.entries());
    this._checkFakeTimers();
    this._immediates.forEach(this._runImmediate, this);
    timerEntries
      .sort(([, left], [, right]) => left.expiry - right.expiry)
      .forEach(([timerHandle, timer]) => {
        this._now = timer.expiry;
        this._runTimerHandle(timerHandle);
      });
  }
  advanceTimersToNextTimer(steps = 1) {
    if (steps < 1) {
      return;
    }
    const nextExpiry = Array.from(this._timers.values()).reduce(
      (minExpiry, timer) => {
        if (minExpiry === null || timer.expiry < minExpiry) return timer.expiry;
        return minExpiry;
      },
      null
    );
    if (nextExpiry !== null) {
      this.advanceTimersByTime(nextExpiry - this._now);
      this.advanceTimersToNextTimer(steps - 1);
    }
  }
  advanceTimersByTime(msToRun) {
    this._checkFakeTimers();
    // Only run a generous number of timers and then bail.
    // This is just to help avoid recursive loops
    let i;
    for (i = 0; i < this._maxLoops; i++) {
      const timerHandleAndExpiry = this._getNextTimerHandleAndExpiry();

      // If there are no more timer handles, stop!
      if (timerHandleAndExpiry === null) {
        break;
      }
      const [timerHandle, nextTimerExpiry] = timerHandleAndExpiry;
      if (this._now + msToRun < nextTimerExpiry) {
        // There are no timers between now and the target we're running to
        break;
      } else {
        msToRun -= nextTimerExpiry - this._now;
        this._now = nextTimerExpiry;
        this._runTimerHandle(timerHandle);
      }
    }

    // Advance the clock by whatever time we still have left to run
    this._now += msToRun;
    if (i === this._maxLoops) {
      throw new Error(
        `Ran ${this._maxLoops} timers, and there are still more! ` +
          "Assuming we've hit an infinite recursion and bailing out..."
      );
    }
  }
  runWithRealTimers(cb) {
    const prevClearImmediate = this._global.clearImmediate;
    const prevClearInterval = this._global.clearInterval;
    const prevClearTimeout = this._global.clearTimeout;
    const prevNextTick = this._global.process.nextTick;
    const prevSetImmediate = this._global.setImmediate;
    const prevSetInterval = this._global.setInterval;
    const prevSetTimeout = this._global.setTimeout;
    this.useRealTimers();
    let cbErr = null;
    let errThrown = false;
    try {
      cb();
    } catch (e) {
      errThrown = true;
      cbErr = e;
    }
    this._global.clearImmediate = prevClearImmediate;
    this._global.clearInterval = prevClearInterval;
    this._global.clearTimeout = prevClearTimeout;
    this._global.process.nextTick = prevNextTick;
    this._global.setImmediate = prevSetImmediate;
    this._global.setInterval = prevSetInterval;
    this._global.setTimeout = prevSetTimeout;
    if (errThrown) {
      throw cbErr;
    }
  }
  useRealTimers() {
    const global = this._global;
    if (typeof global.cancelAnimationFrame === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'cancelAnimationFrame',
        this._timerAPIs.cancelAnimationFrame
      );
    }
    if (typeof global.clearImmediate === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'clearImmediate',
        this._timerAPIs.clearImmediate
      );
    }
    (0, _jestUtil().setGlobal)(
      global,
      'clearInterval',
      this._timerAPIs.clearInterval
    );
    (0, _jestUtil().setGlobal)(
      global,
      'clearTimeout',
      this._timerAPIs.clearTimeout
    );
    if (typeof global.requestAnimationFrame === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'requestAnimationFrame',
        this._timerAPIs.requestAnimationFrame
      );
    }
    if (typeof global.setImmediate === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'setImmediate',
        this._timerAPIs.setImmediate
      );
    }
    (0, _jestUtil().setGlobal)(
      global,
      'setInterval',
      this._timerAPIs.setInterval
    );
    (0, _jestUtil().setGlobal)(
      global,
      'setTimeout',
      this._timerAPIs.setTimeout
    );
    global.process.nextTick = this._timerAPIs.nextTick;
    this._fakingTime = false;
  }
  useFakeTimers() {
    this._createMocks();
    const global = this._global;
    if (typeof global.cancelAnimationFrame === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'cancelAnimationFrame',
        this._fakeTimerAPIs.cancelAnimationFrame
      );
    }
    if (typeof global.clearImmediate === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'clearImmediate',
        this._fakeTimerAPIs.clearImmediate
      );
    }
    (0, _jestUtil().setGlobal)(
      global,
      'clearInterval',
      this._fakeTimerAPIs.clearInterval
    );
    (0, _jestUtil().setGlobal)(
      global,
      'clearTimeout',
      this._fakeTimerAPIs.clearTimeout
    );
    if (typeof global.requestAnimationFrame === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'requestAnimationFrame',
        this._fakeTimerAPIs.requestAnimationFrame
      );
    }
    if (typeof global.setImmediate === 'function') {
      (0, _jestUtil().setGlobal)(
        global,
        'setImmediate',
        this._fakeTimerAPIs.setImmediate
      );
    }
    (0, _jestUtil().setGlobal)(
      global,
      'setInterval',
      this._fakeTimerAPIs.setInterval
    );
    (0, _jestUtil().setGlobal)(
      global,
      'setTimeout',
      this._fakeTimerAPIs.setTimeout
    );
    global.process.nextTick = this._fakeTimerAPIs.nextTick;
    this._fakingTime = true;
  }
  getTimerCount() {
    this._checkFakeTimers();
    return this._timers.size + this._immediates.length + this._ticks.length;
  }
  _checkFakeTimers() {
    if (!this._fakingTime) {
      this._global.console.warn(
        'A function to advance timers was called but the timers APIs are not mocked ' +
          'with fake timers. Call `jest.useFakeTimers({legacyFakeTimers: true})` ' +
          'in this test file or enable fake timers for all tests by setting ' +
          "{'enableGlobally': true, 'legacyFakeTimers': true} in " +
          `Jest configuration file.\nStack Trace:\n${(0,
          _jestMessageUtil().formatStackTrace)(
            new Error().stack,
            this._config,
            {
              noStackTrace: false
            }
          )}`
      );
    }
  }
  _createMocks() {
    const fn = implementation => this._moduleMocker.fn(implementation);
    const promisifiableFakeSetTimeout = fn(this._fakeSetTimeout.bind(this));
    // @ts-expect-error: no index
    promisifiableFakeSetTimeout[_util().promisify.custom] = (delay, arg) =>
      new Promise(resolve => promisifiableFakeSetTimeout(resolve, delay, arg));
    this._fakeTimerAPIs = {
      cancelAnimationFrame: fn(this._fakeClearTimer.bind(this)),
      clearImmediate: fn(this._fakeClearImmediate.bind(this)),
      clearInterval: fn(this._fakeClearTimer.bind(this)),
      clearTimeout: fn(this._fakeClearTimer.bind(this)),
      nextTick: fn(this._fakeNextTick.bind(this)),
      requestAnimationFrame: fn(this._fakeRequestAnimationFrame.bind(this)),
      setImmediate: fn(this._fakeSetImmediate.biÿö} %!À-óÁ¿-ò¿^ôóòÁ-ô	Â‚ À^õôóÅ!
À-óÁ^ôóøÁ-ôÂ‚	À^õôóÁùõhõ«Â„õ
Ã¨öõÀùôhô«  `şÿÿÿ‘•e      Ó€
€€À€   “e    ÎQ    }Q      0T¬´`  md       l   c     4L`   8SbÄâº        !	 !ªa¤ÿ ‚ÿ `şÿÿÿDaÚ^  øa  áİ ò¿U ’Í0Tü`¤  (‘h@      i   x     |      ˆ  	   0L`   8SbÄà¬        á­ ¡­a€ÿ €ÿ I`şÿÿÿDaF_  |`  M•Õ™m0T<´`$   m`    L`    bÌ0TL–`D   —`    $L`    °b         ½C±C á­‘ ¡­µ’¹¡
ya        Dk          ~  )Å¬3ø¬Ä-÷3øø2«  `şÿÿÿ(SbÄà®        I`şÿÿÿDa´_  `  a¤c	       “`  `e     KĞ )   ÏQ      •ia        Dg          -  Æ‚ Ä^ù÷«  `şÿÿÿ(SbÄà®        I`şÿÿÿDa–_  `  —b        ‘e     KĞ (   ÍQ       ²ú0TP´`L   m`     L`    !ª ²É °b         1CéC– á­—¡
¥a        Dl          ÿ¬ Å-ø Æ~)Ä3÷¬3÷^ùø÷«  `şÿÿÿ(SbÄà®        I`şÿÿÿDa*`  x`  a¤c	       L€   `e     KĞ *   ÎQ      Ua        Dà`          … ù%%³ Ã-öÄÂøÿòõ˜!Â]÷ö
¾¡eÍ ñ-ñ˜
-ñ¾ÂñŒ%õ˜!Â]÷ö¾¡eÍ ñ-ñ˜
-ñ¾ÂñŒ%
ÿÀÁŒÀÁ¨¿õ˜#ÿğ-ö ¼]ïö¡»eÍ îŒ½ôšğªò¨ô›óªÂ-õÃ]öõÃ-öÄ‚ Â^÷öõÄ-÷Å‚	Ã^ø÷ö«  `şÿÿÿ’ “f      P @ @ P @ ”e     KĞ '   ÉQ      % ²`0T´`Î   m`    0L`    !ª…İ)U= ²Ì0T,`   ‘`    ]¡
İa        Dc           « `şÿÿÿ(SbÄà®        I`şÿÿÿDa¤a  ¬a  M`e     KĞ ,   ËQ       bÎ0T4´`   m`    L`    ÂÆ—ña        De          -  Æ^ù« `şÿÿÿ(SbÄà®        I`şÿÿÿDaÂa  æa  ‘¤b         “e     KĞ -   ÌQ      ¡
Ía        D|           ¬ ›Œ	Æ-ù Æ!Ä-÷ÅÃ-öÃ^ø÷öÅ-ø
mù›Å-ø
ÆÄ^ùø÷©-Å‚ Ã^øöÅ-øÆ‚	Ä}Ã_ùø÷ö« `şÿÿÿ(SbÄà®        I`şÿÿÿDaº`  òa  M —f      P P ± 
€€&   “e     KĞ +   ÊQ      SbJ €        İ`¤ÿ ’Aa        D@      ùÿÿÿ… ÷%şöõhöÆ%ÿö} %!À-óÁ¿-ò¿^ôóòÁ-ô	Â‚ À^õôóÅ!
À-óÁ^ôóøÁ-ôÂ‚	À^õôóÁùõhõ«Â„õ
Ã¨öõÀùôhô«  `şÿÿÿ’—e      Ó€
€€À€   `e     ÎQ &   ~Q      0T´´`  md       s   ›     (L`   8SbÄáº        ¡­ !«a‚ÿ ‚ÿ `şÿÿÿDa$b  ®c  á¹½U0T<–`$   —`    L`    bÌ0TL–`D   —`    $L`    °b         “C±C ¡­“ !«µ—’¡
I	a        Dk          ~  )Å¬3ø¬Ä-÷3øø2«  `şÿÿÿ(SbÄà®        I`şÿÿÿDa8c  c  1	¤c	       “`  `e     KĞ 0   ÈQ      ‘9	a        Dg          -  Æ‚ Ä^ù÷«  `şÿÿÿ(SbÄà®        I`şÿÿÿDac  šc  “”b        •e     KĞ /   ÇQ      SbJ €        İ`¤ÿ “‘%	a        D˜0      ùÿÿÿ… øş÷öh÷Æ%%ÿ÷š- ˜Âùöhö«- Ã-ö%-Ã/ö%™"Á-ôÂ_õô
Â-õÃ‚ Á^öõôŒÂùöhö«Ã„öÄ¨÷öÁùõhõ«   `şÿÿÿ“”d      P @   •e    ÎQ .   Q      0T´´`  md    "   t   £     4L`   HSbÄâº       !	 á­ !ª ¡­c¤ÿ ¤   $   ‚ÿ !`şÿÿÿDaÌc  ^h  áİ ò¿U ’Í0T	‘`¼  (’h@      i   x     |      ˆ  	   0L`   8SbÄà¬        !« ¡¬a€ÿ €ÿ I`şÿÿÿDa>d  ”e  	•Õ™m!0TP´`L   m`    L`
   İ ¢ÀRcş„í   topicMap bÌ0TL—`D   `    $L`    °b         ½C±C !«“ ¡¬µ”¹¡
½	a        Dk          ~  )Å¬3ø¬Ä-÷3øø2«  `şÿÿÿ(SbÄà®        I`şÿÿÿDaÌd  $e  ¡	¤c	       “`  `e     KĞ 4   ÆQ      —©	a        Dl           !  Ä-÷
