// @google-cloud/firestore/build/src/v1beta1/firestore_client.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirestoreClient = void 0;
const stream_1 = require("stream");
const jsonProtos = require("../../protos/v1beta1.json");
/**
 * Client JSON configuration object, loaded from
 * `src/v1beta1/firestore_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./firestore_client_config.json");
// tslint:disable deprecation
const version = require('../../../package.json').version;
/**
 *  The Cloud Firestore service.
 *
 *  Cloud Firestore is a fast, fully managed, serverless, cloud-native NoSQL
 *  document database that simplifies storing, syncing, and querying data for
 *  your mobile, web, and IoT apps at global scale. Its client libraries provide
 *  live synchronization and offline support, while its security features and
 *  integrations with Firebase and Google Cloud Platform (GCP) accelerate
 *  building truly serverless apps.
 * @class
 * @deprecated Use v1/firestore_client instead.
 * @memberof v1beta1
 */
class FirestoreClient {
    /**
     * Construct an instance of FirestoreClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://developers.google.com/identity/protocols/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean | "rest"} [options.fallback] - Use HTTP fallback mode.
     *     Pass "rest" to use HTTP/1.1 REST API instead of gRPC.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new FirestoreClient({fallback: 'rest'}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        var _a, _b;
        this._terminated = false;
        this.descriptors = {
            page: {},
            stream: {},
            longrunning: {},
            batching: {},
        };
        // Ensure that options include all the required fields.
        const staticMembers = this.constructor;
        const servicePath = (opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint) || staticMembers.servicePath;
        this._providedCustomServicePath = !!((opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint));
        const port = (opts === null || opts === void 0 ? void 0 : opts.port) || staticMembers.port;
        const clientConfig = (_a = opts === null || opts === void 0 ? void 0 : opts.clientConfig) !== null && _a !== void 0 ? _a : {};
        const fallback = (_b = opts === null || opts === void 0 ? void 0 : opts.fallback) !== null && _b !== void 0 ? _b : (typeof window !== 'undefined' && typeof (window === null || window === void 0 ? void 0 : window.fetch) === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== staticMembers.servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = staticMembers.servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === staticMembers.servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        // Determine the client header string.
        const clientHeader = [`gax/${this._gaxModule.version}`, `gapic/${version}`];
        if (typeof process !== 'undefined' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else if (opts.fallback === 'rest') {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = {
            listDocuments: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'documents'),
            partitionQuery: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'partitions'),
            listCollectionIds: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'collectionIds'),
        };
        // Some of the methods on this service provide streaming responses.
        // Provide descriptors for these.
        this.descriptors.stream = {
            batchGetDocuments: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, opts.fallback === 'rest'),
            runQuery: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.SERVER_STREAMING, opts.fallback === 'rest'),
            write: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.BIDI_STREAMING, opts.fallback === 'rest'),
            listen: new this._gaxModule.StreamDescriptor(this._gaxModule.StreamType.BIDI_STREAMING, opts.fallback === 'rest'),
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.firestore.v1beta1.Firestore', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.firestoreStub) {
            return this.firestoreStub;
        }
        // Put together the "service stub" for
        // google.firestore.v1beta1.Firestore.
        this.firestoreStub = this._gaxGrpc.createStub(this._opts.fallback
            ? this._protos.lookupService('google.firestore.v1beta1.Firestore')
            : // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this._protos.google.firestore.v1beta1.Firestore, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const firestoreStubMethods = [
            'getDocument',
            'listDocuments',
            'updateDocument',
            'deleteDocument',
            'batchGetDocuments',
            'beginTransaction',
            'commit',
            'rollback',
            'runQuery',
            'partitionQuery',
            'write',
            'listen',
            'listCollectionIds',
            'batchWrite',
            'createDocument',
        ];
        for (const methodName of firestoreStubMethods) {
            const callPromise = this.firestoreStub.then(stub => (...args) => {
                if (this._terminated) {
                    if (methodName in this.descriptors.stream) {
                        const stream = new stream_1.PassThrough();
                        setImmediate(() => {
                            stream.emit('error', new this._gaxModule.GoogleError('The client has already been closed.'));
                        });
                        return stream;
                    }
                    return Promise.reject('The clH\ HLHH C DA@ uHOHtHLHH nC DAH _^[HOH[ HW H[ HJHHO HuE3AH _^[HtLID  3LE3@UHH0MLQxM[ DA8tpA|N   H`DEBfEBfEBEB	EB
EBEBEB
EBEBEMHU  H0]ËD 9BuD 9BuD 9BxLIhMkIMIH *C 봋~D 9BHsD 9B9hD 9B*IHH@C 3pH\$Ht$WH HIH+F HHuHBH+	F H7  HH+F uHBH+F H  H`   HEHH@0oC HH  HH+F uHCH+F HZ HH+F uHCH+F HZ HH+iF uHCH+dF HZ HH+YF uHCH+TF HZ HH+IF uHCH+DF HHZ HLHH C DA@ uHOHtHLHH C DAH\$0Ht$8H _Ht;LIA  3LHOHY HA  H\$0E3Ht$8AH _E3@UHH@LLHIxH,Z HED I      H>D I      IJ`AEAAfEAAfEAAEAA	EAA
EAAEAAEAA
EAAEAAEHUH@]IJhHpHIH NC MIIH@9C 3̋9t3ËB9AuB9AuB9Au   @SVWH HIH+;F HHuHBH+0F HZ HH+%F uHBH+ F HY H   HEHH@0C HHt8HOHY HW HY HJHHO Hu3ɋH _^[HH+܁F uHCH+ׁF HX HH+F uHCH+F HX HH+\F uHCH+WF HX HH+LF uHCH+GF HX HH+<F uHCH+7F HHX HLHH C ȁ@ 3LOM&ILHIH qC 
@SVWH HIHIHHX HHH@02C HH   HH+F uHCH+F HB  HH+|F uHCH+wF H[X HH+,F uHCH+'F H   HH+F uHCH+F H   HH+F uHCH+F HH   HLHH qC Ё@ uBHOHt9HLHH LC #HOHW HW Hy>   D3ҋH _^[HH@8HC HH O 3ҋH _^[HJHHO Ht/?  3HH EUHH@@SVWH HIH+~F HHuHBH+~F Hi  HH+~F uHBH+~F HI  HH   HEHH@0FC HH   HH+~F uHCH+~F HV HH+~F uHCH+~F HV HH+@~F uHCH+;~F HV HH+0~F uHCH++~F HsV HH+ ~F uHCH+~F HHV HLHH C DA@ uHOHtHLHH ^C DAH _^[HOHU HW HU HJHHO HuE3AH _^[HtLI;  3LE3@SVWH HIHIHHV HHH@0C HH   HH+<}F uHCH+7}F HV HH+}F uHCH+}F HV HH+|F uHCH+|F HU HH+|F uHCH+|F HU HH+|F uHCH+|F HHU HLHH C ȁ@ u LOMtILHIH C ȋH _^[HOHU HW HU HJHHO Hu3ɋH _^[@SVWH HIH+{F HHuHBH+{F Hi  HH+{F uHBH+{F HI  HX   HEHH@0&C HH   HH+{F uHCH+{F HT HH+p{F uHCH+k{F HT HH+ {F uHCH+{F HT HH+{F uHCH+{F HT HH+ {F uHCH+zF HHT HLHH e
C DA@ uHOHtHLHH >
C DAH _^[HOHS HW HT HJHHO HuE3AH _^[HtLIx8  3LE3@SVWH HIH+zF HHuHBH+ zF Hi  HH+yF uHBH+yF HI  Hh   HEHH@0fC HH   HH+yF uHCH+yF HS HH+yF uHCH+yF HS HH+`yF uHCH+[yF HS HH+PyF uHCH+KyF H{S HH+@yF uHCH+;yF HHS HLHH C DA@ uHOHtHLHH ~C DAH _^[HOHR HW HR HJHHO HuE3AH _^[HtLI6  3LE3@SVWH HIH+KxF HHuHBH+@xF Hi  HH+5xF uHBH+0xF HI  H@   HEHH@0
C HH   HH+ xF uHCH+xF HR HH+wF uHCH+wF HR HH+wF uHCH+wF HR HH+wF uHCH+wF HaR HH+wF uHCH+{wF HHR HLHH 	C DA@ uHOHtHLHH 	C DAH _^[HOHQ HW HQ HJHHO HuE3AH _^[HtLI4  3LE3@SVWH HIH+vF HHuHBH+vF Hi  HH+uvF uHBH+pvF HI  H    HEHH@0C HH   HH+`vF uHCH+[vF HQ HH+0vF uHCH++vF HQ HH+uF uHCH+uF HQ HH+uF uHCH+uF HGQ HH+uF uHCH+uF HHQ HLHH %C DA@ uHOHtHLHH C DAH _^[HOHP HW HP HJHHO HuE3AH _^[HtLI83  3LE3@SVWH HIH+tF HHuHBH+tF Hi  HH+tF uHBH+tF HI  HP   HEHH@0&C HH   HH+tF uHCH+tF HP HH+ptF uHCH+ktF HP HH+ tF uHCH+tF HP HH+tF uHCH+tF H-P HH+ tF uHCH+sF HHqP HLHH eC DA@ uHOHtHLHH >C DAH _^[HOHO HW HO HJHHO HuE3AH _^[HtLIx1  3LE3@SVWH HIH+sF HHuHBH+ sF H+  HH+rF uHBH+rF H  H`   HEHH@0fC HH  HH+rF uHCH+rF HcO HH+rF uHCH+rF HeO HH+`rF uHCH+[rF HfO HH+PrF uHCH+KrF HO HH+@rF uHCH+;rF HHEO HLHH C Ё@ uHOHtHLHH C ЋH _^[HteHQLA MO    CD H3HOHTN HW Hy   D3ҋH _^[HJHHO Ht3@UHH0MLQ`MN DA8tpA|N   HHDEBfEBfEBEB	EB
EBEBEB
EBEBEMHU   H0]ËC 9BuC 9BuC 9BxLIPMkIMIH *C 봋~C 9BHsC 9B9hC 9B*IHH@C 3p@SVWH HIH+pF HHuHBH+pF Hg  HH+pF uHBH+ pF HG  HH   HEHH@0vC HHt:HOHM HW HM HJHHO HuE3AH _^[HH+oF uHCH+oF HqM HH+oF uHCH+oF HtM HH+:oF uHCH+5oF HuM HH+*oF uHCH+%oF HM HH+oF uHCH+oF HHUM HLHH C DA@ 0HOH#HLHH PC D	HtLI,  3LE3H\$Ht$WH HIH+TnF HHuHBH+InF HV  HH+>nF uHBH+9nF H6  HH@0 C HH   HH+5nF uHCH+0nF HL HH+nF uHCH+ nF HL HH+mF uHCH+mF HKL HH+mF uHCH+mF H"L HH+mF uHCH+mF HH\L HLHH B DA@ uHOHtHLHH B DAH\$0Ht$8H _HOHK H)  H\$0E3Ht$8AH _HtLI)  3LE3@SVWH HIHHH@0RB HH   HH+lF uHCH+lF HK HH+lF uHCH+lF HK HH+LlF uHCH+GlF HteHH+@lF uHCH+;lF HK HH+0lF uHCH++lF HHK HLHH H _^[H%B H _^[HHH@8uB HHGHx-HHHOu3H _^[HGHxHHHOt   L @SVWH HIH+[kF HHuHBH+PkF HO  HH+EkF uHBH+@kF H/  HH@0B HHt1HOHJ HG H5     L 3ҋH _^[HH+kF uHCH+
kF HJ HH+jF uHCH+jF HJ HH+jF uHCH+jF HJ HH+jF uHCH+zjF H,J HH+ojF uHCH+jjF HHnJ HLHH B Ё@ 3HOH&HLHH B 
HtEHQHA HxHHHO uI3   L HHHO 3@VH@H[xZ H3HD$8H	HHH;   H   H   HLD$0H\$`HciF 3H|$hH\$0H B HH|$0H|$ H\$(Ht"HLD$(H(iF H B H\$(H\$(H;t[@2HH\$`t
HL$() HH|$ht
HL$  @HL$8H3j
 H@^ðHL$8H3oj
 H@^2@@SVWH HIHHH@02B HH   HH+hF uHCH+hF HH HH+|hF uHCH+whF HH HH+,hF uHCH+'hF H   HH+hF uHCH+hF HtVHH+hF uHCH+hF HH{H HLHH uB HGHxjHHHOu3H _^[HHH@DB HHGHx8HHHOtHHH@8B HHGHxHHHOt   L @SVWH HIHHH@0B HH   HH+LgF uHCH+GgF H   HH+gF uHCH+gF HG HH+fF uHCH+fF HG HH+fF uHCH+fF Ht7HH+fF uHCH+fF HH   HLHH B .HHH@B HHGHQG HHHOu3H _^[HGH.G HHHOtHHH@B HHGH G HHHOt'  HH3Ҹ HEH _^[HH+eF uHBH+eF 3HtHHAHHH#@SVWH HIHHH@0B HH   HH+eF uHCH+eF HTF HH+\eF uHCH+WeF HeF HH+eF uHCH+eF HfF HH+dF uHCH+dF HsF HH+dF uHCH+dF HHF HLHH H _^[H%IB H _^[HGHNF HHHOu3H _^[HAHHH#@SVWH HIHHH@0B HH   HH+LdF uHCH+GdF HE HH+dF uHCH+dF HE HH+cF uHCH+cF H   HH+cF uHCH+cF HE HH+cF uHCH+cF HHE HLHH H _^[H%	B H _^[HGHx@HHHOu3H _^[HHH@8B HHGHxHHHOt   L H\$Ht$WH HIH+bF HHuHBH+bF Hb  HH+bF uHBH+bF HB  HX   HEHH@0/B HH   HH+bF uHCH+bF HD HH+ybF uHCH+tbF HD HH+)bF uHCH+$bF H E HH+bF uHCH+bF HD HH+	bF uHCH+bF HHD HLHH nB DA@ uHOHtHLHH GB DAH\$0Ht$8H _HOH
D H  H\$0E3Ht$8AH _HtLI  3LE3@UHH0MLQxMGD DA8tpA|N   H`DEBfEBfEBEB	EB
EBEBEB
EBEBEMHUH0]ËC 9BuC 9BuC 9BxLIhMkIMIH 
B 봋^C 9BHSC 9B9HC 9B*IHH@B 3p@UHH0ML   MC DA8D A|NtcHhDEBfEBfEBEB	EB
EBEBEB
EBEBEMHUbH0]ËC 9BuzC 9BusC 9BxIHH@B 3@SVWH HIH++_F HHuHBH+ _F H+  HH+_F uHBH+_F H  H    HEHH@0B HH  HH+ _F uHCH+^F H-C HH+^F uHCH+^F H/C HH+^F uHCH+{^F H0C HH+p^F uHCH+k^F HB HH+`^F uHCH+[^F HHC HLHH B Ё@ uHOHtHLHH B ЋH _^[HtKHQLA MB IHIHO LuH3HOHB H  3ҋH _^[3H\$Ht$WH IHHt>HYH=@ uHKHtHLHH B H\$0Ht$8H _3@UHH0MLQpM+B DA8tpA|N   HXDEBfEBfEBEB	EB
EBEBEB
EBEBEMHUH0]ËC 9BuC 9BuC 9BxLI`MkIMIH 
B 봋^C 9BHSC 9B9HC 9B*IHH@B 3pH\$Ht$WH HIHIHHuA HHH@0B HH&  HH+\F uHCH+ \F H^A HH+[F uHCH+[F H`A HH+[F uHCH+[F H   HH+u[F uHCH+p[F HtnHH+i[F uHCH+d[F HH5A HLHH B Ё@ uHOHtHLHH B ЋH\$0Ht$8H _HH@HB HHOH@ H  H\$03Ht$8H _HOHt7@ HH@8HAHHH#>H\$Ht$WH HIH+TZF HHuHBH+IZF H@ HH+>ZF uHBH+9ZF H@ HH@0B HH   HH+5ZF uHCH+0ZF H@ HH+ZF uHCH+ ZF H@ HH+YF uHCH+YF H@ HH+YF uHCH+YF H? HH+YF uHCH+YF HH? HLHH B Ё@ uHOHtHLHH B ЋH\$0Ht$8H _HOH&? H<  H\$03Ht$8H _HAHHH#H\$Ht$WH HIH+XF HHuHBH+XF H7  HH+XF uHBH+XF H  H    HEHH@0B HH  HH+yXF uHCH+tXF H? HH+IXF uHCH+DXF H$? HH+WF uHCH+WF H> HH+WF uHCH+WF H> HH+WF uHCH+WF HH> HLHH >B DA@ uHOHtHLHH B DAH\$0Ht$8H _Ht;LI  3LHOH> Hg  H\$0E3Ht$8AH _E3@SVWH HIH+VF HHuHBH+VF HT  HH+VF uHBH+VF H4  X   HHDHH@0RB HH   HH+VF uHGH+VF H= HH+VF uHGH+VF H> HH+LVF uHGH+GVF H= HH+<VF uHGH+7VF H= HH+,VF uHGH+'VF HH= HLHH B =@ uHKHtHLHH oB H _^[HKH= HS H= HJHHK Hu3  HHHH#HH\$UHH0MLY03M;= DA8= A|N= HDEBfEBfEBEB	EB
EBEBEB
EBEBEMHUs؋H\$@H0]H\$Ht$WH HIHIHH= HHH@0KB HH   HH+TF uHCH+TF H4  HH+TF uHCH+TF Hi= HH+ETF uHCH+@TF H   HH+5TF uHCH+0TF H   HH+%TF uHCH+ TF HH   HLHH B Ё@ uHOHtHLHH eB ЋH\$0Ht$8H _HOH< HL  H\$03Ht$8H _HH@8HB HHOHt^< V  3HH EcHH@H\$Ht$WH HIHIHHG< HHH@0B HH   HH+%SF uHCH+ SF H0< HH+RF uHCH+RF H;< HH+RF uHCH+RF H; HH+RF uHCH+RF H; HH+RF uHCH+RF HH; HLHH B Ё@ uHOHtHLHH B ЋH\$0Ht$8H _HOHB; H,  H\$03Ht$8H _HAHHH#>@SVWH HIHHH@0BB HH   HH+QF uHCH+QF H0; HH+QF uHCH+QF HA; HH+<QF uHCH+7QF HtvHH+0QF uHCH++QF H   HH+ QF uHCH+QF HH; HLHH B HGHxjHHHOu3H _^[HHH@8TB HHGHx8HHHOtHHH@*B HHGHxHHHOt   L HAHHH#H\$Ht$ UWATAVAWHl$H   H^Z H3HE/LLE3DeLeH
HHUH@0B : H}H}H ~Z H]HtHHH@_B H;  H9 H9 LeHLEHOF HH &B H}H}LeHLEHlOF HH  B HEHEH;9 @2Ht	HMnw Ht	HM`w Ht	HMRw @  H
}Z HMHt
HH@B A   HUHM:  H
;}Z HMHt
HH@eB C   HUHM  @2H} t	HMv H} t	HMv @8 H
}Z HMHt
HH@B HUHMv  H
|Z HMHt
HH@B HUHMgA  @2tH} t	HM$v H} t	HMv @  H
{Z HMHt
HH@[B HUHMH} t	HMu @7 H
+{Z HMHt
HH@B HUHMH} t	HMu @7 H
}{Z HMHt
HH@B HUHMrH} t	HM?u @7 H
{Z HMHt
HH@B HUHM,H} t	HMt @  H
zZ HMHt
HH@CB HUHMH} t	HMt @  H
zZ HMHt
HH@B HUHMH} t	HMmt @R  H
zZ HMHt
HH@B HUHMZH} t	HM't @  H
yZ HMHt
HH@qB HUHMH} t	HMs @  H
yZ HMHt
HH@+B HUHMH} t	HMs @6 H
CyZ HMHt
HH@B HUHMH} t	HMUs @6 H
yZ HMHt
HH@B HUHMBH} t	HMs @6 H
wxZ HMHt
HH@YB HUHMH} t	HMr @n7 H
xZ HMHt
HH@B HUHMH} t	HMr @67 H
sxZ HMHt
HH@B HUHMpH} t	HM=r @7 H
uxZ HMHt
HH@B HUHM*H} t	HMq @6 H
wZ HMHt
HH@AB HUHMH} t	HMq @6 H
QwZ HMHt
HH@B HUHM6 H
wZ HMHt
HH@B  HUHMfh6 @2 tH} t	HM&q H} t	HMq @?6 H
vZ HMHt
HH@`B HUHMH} t	HMp 3 H
vZ HMHt
HH@B HUHMH} t	HMp 5 H} t	HMsp HM/H3I
 L$   I[@IsHIA_A^A\_]@I  I   a  EHUI,  뛲IP  AAIc {@IR	 fAƇ   HUIIHpI肘 H} t	HMo IIf	 HUI譋H8HUIH!DËId	 H} t
HM_o H} t	HMNo I  I&I I@ESH\$UHH0MLYp3MK4 DA84 A|NtcHXDEBfEBfEBEB	EB
EBEBEB
EBEBEMHUW؋H\$@H0]Ë>C 9Bu7C 9Bu0C 9BxI
HH@B H\$UHH0MLY83M/4 DA8tpA|N   H DEBfEBfEBEB	EB
EBEBEB
EBEBEMHU7؋H\$@H0]Ë^C 9BuWC 9BuPC 9BxLI(MkIMIH B 뫋(C 9BHC 9B9C 9B*I
HH@B kH\$UHH0MLYp3Mq3 DA83 A|NtcHXDEBfEBfEBEB	EB
EBEBEB
EBEBEMHUK؋H\$@H0]ËNC 9BuGC 9Bu@C 9BxI
HH@B H\$UHH0MLY83MU3 DA83 A|NtcH DEBfEBfEBEB	EB
EBEBEB
EBEBEMHU؋H\$@H0]Ë~C 9BuwC 9BupC 9BxI
HH@B HQ HxLBHLA HuAø   DH\$UHH0MLY(3M	3 DA8]3 A|N3 HDEBfEBfEBEB	EB
EBEBEB
EBEBEMHU؋H\$@H0]H\$UHH0LHAHML#HDC IĿ3ۅJ3 A|N]3 IJ UAAfEAAfEAAEAA	EAA
EAAEAAEAA
EAAEAAEHU؋H\$@H0]HQ H93 LBHLA HuAHAHHH#HQH
3 H\$Ht$WH HIHIHH   HHH@0B HH   HH+eAF uHGH+`AF H2 HH+5AF uHGH+0AF H2 HH+@F uHGH+@F H   HH+@F uHGH+@F Hh2 HH+@F uHGH+@F HH2 HLHH *B =@ uHKHtLHHH B H\$0Ht$8H _HHH@8B HHKH1 H`3HAHHH#H\$Ht$H|$ AVH0HNZ H3HD$(Hy LHy6   t?H}   HL$(H3lA
 H\$HHt$PH|$XH0A^HD     A
 Hؾ   HtvIIH@B 
&}Z HC HSH
dC H{sHCH\$ Ht7HH       HHHIN H$1 sHCI33>H    HHHDHX   HHHD@UHH0LLHC I0 HC I0 AEABfEABfEABEAB	EAB
EABEABEAB
EABEABEIIXHUH0]HAHHH#FH\$Ht$H|$ AVH0HLZ H3HD$(HyLHR0 H٪D     ?
 Hؾ   H   IIH@FB 
T{Z HC HSH
C H{sHCH\$ HtUHH       HHHINH/ sHCHL$(H3>
 H\$HHt$PH|$XH0A^33HAHHH#   H\$H|$UHH0IDA8/ A|N/ HY DEBfEBfEBEB	EB
EBEBEB
EBEBEHEHH+
g<F uHMH+
b<F Ht\H+^<F uHEH+Y<F HtCLHUHQ   =@ uHKHtHHULH B H\$@H|$HH0]H1HKHHH#H3H\$Ht$WH Hɸ    HIHDHHH@0[B HH@/ HH+;F uHGH+;F H
/ HH+;F uHGH+;F H. HH+U;F uHGH+P;F H. HH+E;F uHGH+@;F H. HH+5;F uHGH+0;F HH. HLHH B H\$0Ht$8H _H\$Hl$VWAVH@)t$0HIZ H3HD$(E(H+    H      H        HΉ|$ 2    ;   tK. 3|$ EuD8  t   ƃ   HT  EHT$ HˋT  |$ tr. uH HL$(H3;
 H\$hHl$p(t$0H@A^_^Hz	 f/sf/5E i-    VH蚞 dH  t   uH8	 hHHXHpHx UHH`)pHHZ H3HEHH   H2    
   V- H<  H  H	  H  H|  H     薘]- H      H  h  HUH  |HEHtFHUHM   uH1 uoWEEuMHUHM   HEHt	HM` HMH39
 L\$`I[Is I{((t$PI]À   7, t, @SH@HFZ H3HD$0H3HBH	Ht"H!D$(HLD$(H8F H oB HL$(HL$ HHH@PUB x"HL$ _ HHL$0H3G9
 H@[ŋL& @SHPH/FZ H3HD$HHH	Ht#Hd$@ HLD$@H7F H B HL$@HL$ HD$0ST$8HT$0H@XB xHL$ ;_ HL$HH38
 HP[뵋% H3   @SH0HEZ H3HD$(LHHt?HLD$ Hd$  H6F IH 3B HD$ HHHL$(H3+8
 H0[H! Hʲ   @SH@HDZ H3HD$0H	Ht#Hd$( HLD$(H`6F H B HL$(HL$ HH@8B xHL$  ^ HL$0H37
 H@[ɋ$ H\$Ht$WH AHH{  t5HU  HM  HyM  HH"   H\$0Ht$8H _H   JO  ;tH\$Ht$ UWAVHH`HDZ H3HEHHHMBHLcZ LEHM' LHUHH} t
HM.] H} t	HM]     7
 HH  H` H` LpH5F IA   A
krZ LBb HC HHUȋBWEH   H+) H   H]H}H$) H) HM#   HMH35
 L\$`I[0Is8IA^_]@SH HH! H7u HHCHtHHH@B H [H\$WH HH
qZ @^t
    H5
 H\$0HH _H\$UVWAVAWHl$H   HIBZ H3HE'MHL@  IH( H  H]HM!& HtEH  H]HM& HtHM  tPHM  tBHt
HMC[ H} t	HM2[ HM'H34
 H$   HĐ   A_A^_^]A   HHD HM蕷 HOhHtgHe HLEH5F H DB HEHEEHEHEHEHEHEHEHMs  H} t	HMZ HBJ  8He H\$H|$UHH`H	      =^Z     ^Z    e H  H]HtHHH@B HtH   + ueHt	HMY H`  H]HtHHH@KB HtHUHM   H]Ht	HMY L\$`I[I{ I]HM  tnHM   t`H]wuJA'   HC HM    fEHEHEHEHEHMKr]Z j]Z E   @SH@H?Z H3HD$0HH	Ht#Hd$( HLD$(HO=F H NB HL$(HL$ HH   2B xHL$ X HL$0H3'2
 H@[Ƌ, H8H?Z H3HD$(H	HT$ d$  HH   B xD$ HL$(H31
 H8Ë @SHp
hWZ eH%X      Hȋ
9wZ ^\Z Hp[H
vZ 3
 =vZ u280\Z    susA'   HC HL$0d 	   f$   HD$0HD$ H$   HD$(HL$ 覑[Z [Z /% [Z H
^vZ 2
 UΊ[Z H\$H|$UHl$H   H=Z H3HEOHH   H}HM! H)  HJ uH   ( :  HcC HEHE   HUHMGv  HM?@  HEGHEHUHM,HHM?U  H} t	HMV He7 HLE7HQ?F HH B HE7HEHEHEHE?HEHmZ HEHzmZ H
kmZ H2  HLE?HUH@pB $ H
EmZ H} t
HM
V H}? t
HM?U HMGv H} t	HMU HMOH3U/
 L$   I[I{ I]He He/ HLE/Hp>F HH B HE/HEHEHEHEHEHlZ HEHlZ H
lZ HtBHLEHUH@pB xRH
llZ H} t
HM4U H} 7
# H
ClZ HU)	 H
0lZ HU)	  H\$Ht$H|$ AVH0H\;Z H3HD$(Hy LH" HD     .
 Hؾ   H   IIH@B 
jZ HqC HSH
FC H{sHCH\$ HtUHH       HHHIN HL" sHCHL$(H3-
 H\$HHt$PH|$XH0A^33@SHPHHT$Hd$  HH " HD$(H# D$    H HH@'B xHL$(GHHP[Ë- H\$WH`H:Z H3HD$PHL$ IHD HA   HL$( LD$(HHT$ HL$ H! H!HHL$PH3,
 H\$xH`_H\$UHH`H9Z H3HEHHMH! He He HEHEHEHEH}jZ HEHqjZ H=ajZ  ubH
_jZ LEHU HUH>r H} C! H} t
HMR H} 6! HHMH3+
 H\$xH`]LMLEHUH
iZ -H
iZ @SH@H8Z H3HD$0HH	Ht#Hd$( HLD$(H[6F H jB HL$(HL$ HHH  MB xHL$ Q HL$0H3B+
 H@[ŋG H8H-8Z H3HD$(H	HT$ d$  HH@8B xD$ HL$(H3*
 H8Ë @SH@H7Z H3HD$8HHT$(Hd$( H	Ht#Hd$0 HLD$0H2F H B HL$0HL$ HHT$(H@0kB x*HL$ P HL$(HHHL$8H3U*
 H@[뻋Z H\$UHH`H87Z H3HEHHMH! He He HEHEHEHEHEhZ HEH9hZ H=)hZ  ubH
'hZ LEHU HUHo H}  H} t
HM+P H}  HHMH3)
 H\$xH`]LMLEHUH
gZ +H
gZ @SH@H[6Z H3HD$0HH	Ht#Hd$( HLD$(H(F H B HL$(HL$ HHH@hB xHL$ zO HL$0H3(
 H@[ȋ H(H	HHH@`B xH(Ë @SH@H5Z H3HD$0HH	Ht#Hd$( HLD$(HW3F H fB HL$(HL$ HH   JB xHL$ N HL$0H3?(
 H@[ƋD HX)t$@H$5Z H3HD$0(H	Ht#Hd$( HLD$(H2F H ۸B HL$(HL$ H(H   B x#HL$ @N HL$0H3'
 (t$@HX @SH`H4Z H3HD$XHH	Ht#Hd$P HLD$PH?2F H NB HL$PHL$ H)D$0K)L$@HT$0H   B xHL$ M HL$XH3'
 H`[벋 @SH@H3Z H3HD$0HWBH	Ht#Hd$( HLD$(H1F H B HL$(HL$ HHH   B x"HL$ 
M HHL$0H3}&
 H@[ @SH HHL$0!  H\$0Ht"H8   Ht
HL$0L H [3LI[WH@HISIc HH! IC  HT$0HL$ 5   HHHB H|$  W H|$( [ H\$XH@_@SH0H2Z H3HD$(H	HH: H!
HHL$(H3u%
 H0[HT$USWHH@HH  H3H   HU k  HE H   HHM(  H}( tqHh HEe (EfELEHHM0=  LE0HU8HM(   L HU(HM}
 HUH_  HM  H}0 t
HM0)K H}( t	HM(K HM _e HE Ht	HM J H   HH@_[]`  @SH H" MH	HLIHH@0&B x	HH [Ë6 @SH0H1Z H3HD$(HL$ LHHtCHLD$ Hd$  H5F IH ʴB x$HD$ HHHL$(H3#
 H0[H!  @USVWHHHHH  H](HM( Hi H}( & HH_^[]LSH@H]0Z H3HD$8Ic HH
H" HBHIKIKICV HT$0HI H|$0  HL$(HSHKHL$ HT$(   HHL$8H3"
 H@[H\$WH0H9 H H\$HH0_@SHPHHL$ )t$@(   A0HT$(t$0#
 D$(   HD$ f@p
k^Z (t$@H
/C P HHHHP[H\$HT$WH@H2/Z H3HD$0IIHtI  Hd$( I! LD$(HT$XHL$    HD$ HHt2HL$(3HD$(HH 3HL$0H3!
 H\$PH@_HD$(֋D$(@SH0H.Z H3HD$(HL$ LHHtCHLD$ Hd$  H1F IH RB x$HD$ HHHL$(H3F!
 H0[H! G H\$VWAVH0MHHHL$ `  !
 HHD$ 3A`  HJ
 H   HC HHڭC HC`HǭC H   HHCxHKpHL$ HT$ If H|$   IHHH\$XH0A^_^H\$WH@HHyB HSx   HHD; HHH#	  3H\$XH@_ËD$PHL$SVWH HYH7C HHC HG`H   H
C HH      D@H   H   3H   H   H   H   H   H   H   H   H   H  H  H  H   H(  H0  H8  H@  HH  HP  HX  HL$PHH
~LZ HL$HHL$H; LHT$HHH9t$HtHL$HE H9t$Pt
HL$PrE H   HH _^[H\$UHHPH+Z H3HEHH`HHUH@(B HUHEHKh3HtaH]HLEHF H fB HEHEHUHMeH9]t
HMD HMe HMH3;
 H\$hHP]H]H\$Ht$UWAVHl$Hp  H+Z H3HEhH8E3]  HNhH: LuHLEHF H B HEHD$0LD$0HL$(L9t$0t
HL$0D L9t$(  HMHHL$(c L9u A   HE HMHv HUHHL$(HNhH LuHLEH5F H B HEHD$ LD$ HWD HM *=  L9t$ t
HL$ tC L9u   HT$`HM \HHL$(H@ LuHLEH'F H B HEHD$8](HD$8HEHE(HEH=ZZ H}HZZ H
qZZ H  H
hZZ HU	 L9t$8tHL$8B L9t$`t
HL$`B HL$(H Lu HLE H&F H B HE HD$@Du0HD$@HEHE0HEH}HYZ H
YZ Hi HDE0HT$@H@@B C H
YZ L9t$@t
HL$@B HT$hHM ^HRL9t$ht
HL$hA a  HL$(H LuHLEH&F H +B HEHD$HDu8HD$HHEHE8HEH}HYZ H
YZ H HDE8HT$HH@X۫B  H
XZ L9t$Ht
HL$HJA HT$pHM ]HHL$(H| LuHLEHi%F H xB HEHD$P]@HD$PHEHE@HEH}HgXZ H
XXZ H: HDE@HT$PH   &B  H
.XZ L9t$PtHL$P@ L9t$pt
HL$p@ HUHM HH\$(H LuHLEH)F HH B HEHD$ HT$ H5L9t$ tHL$ @ L9ut	HM@ HL$XAHUHL$X.YHHv LuHLEH(F HH <B HEHD$ HT$ HL9t$ tHL$ ? L9ut	HM? H  H\$0HL$0
 H   HT$xHL$0u[H\$xH   LuHLEHF HH B HEHEHtSHL$XH LuHLEHF H rB HEHD$ HT$ HML9t$ tHL$ > L9ut
HM> HtHL$x> L9t$0tHL$0> L9t$XtHL$X> L9u t	HM > (~E EH(
E MXDt$0HT$0HL$('H   /  
  t\
ZE HL$(HUHHL$(L9t$(t
HL$(> HMhH3
 L$p  I[(Is0IA^_]HDE(HT$8H@pFB x
H
RUZ R H\$Ht$UWAVHHPH($Z H3HEHHIhL5F 3H" HuHLEIH ٧B HEHEHOhH HuHLEIH B HEHELEHC HMm  HH   	H9ut
HM= H9ut	HM< HOhH HuHLEIH =B HEHELEH*D HMU6  H9ut	HM< HEHtH   HUHEHt	HMv< LEHD HMHH   JH9ut	HMB< H  HHtbHn  
 3HJ  H   HU  H9ut"HM  tA   H(H9ut	HM; HOhH
 HuHLEIH B HEHELEHD HM$  HH0     H9uI
 H9ut	HMf; HOhH<
 HuHLEIH B HEHELEHD HM   HHH  V   H9u H9ut	HM: H  H9ut	HM: HMH3X
 L\$PI[(Is0IA^_]H\$Ht$WH HHHIHy  HHsHHH H9y H\$0Ht$8H _H\$H|$ UHH`H Z H3HEHMHًJ3L
tnfA9<IulE   HUЉMLMILEHUH}HH@@tB xAHMHMH H;HHMH3]
 L\$`I[I{(I]HHB ̋L  H\$H|$ UHH`H# Z H3HEHMHًJ3L
 fA9<I E   HUЉMLMILEHUH}HH@@B x^HMHMHtLHLEH}HH!F B HMHHM9 HHMH3y
 L\$`I[I{(I]H;ڋo
 H\$H|$UHH@H?Z H3HEHH   HU9   HH} t	HM8 H
 HMH3
 H\$XH|$`H@]@SH0HZ H3HD$(H	HHt=Hd$  LD$ HHG!F H@B HD$ HHHL$(H3
 H0[H" H\$H|$UHHpHoZ H3HEHH   H]Hu(Hu5HuoHMH37
 L\$pI[I{ I]HHH@B HUHM(  EMMMMHUHM>H]HMK7 HHHZ H3HD$8d$0 H	Ht#Hd$( HLD$(H; F H zB HL$(HL$ HHT$0H   [B x"HL$ 6 D$0HL$8H3L
 HHR
 H\$H|$UHH`H+Z H3HEe He E   H   HUH)     ]HUH脤u?HUH     ]e HHHUH@8B 	 } 	 He ]HUH  ]HUH H   @2tH}  tH}  tH} t
HM5 tH} t
HM5 tH}  @HMH3
 L\$`I[I{ I]H\$UVWHHPHZ H3HEHHHUH;Z H]HtHHH@yB He HFh   HHDHHt He HLEHF H ?B HMHMHLEHH@0#B  HM4 HuHuHtfHe HLEH:F HH B xIHEHHt
HM]4 Ht	HMN4 HHMH3
 H$   HP_^]H' 뾋
 H\$H|$UHHPHZ H3HEHH:Z H]HtHHH@MB He HGh   HHDHHt He HLEHtF H B HMHMHLEHH@0B x[HMz3 HEHEHM @H} t
HMV3 Ht	HMG3 @HMH3
 H\$hH|$pHP]Ë
 H\$UVWHHPHZ H3HEHHHUHj9Z H]HtHHH@IB He HFh   HHDHHt He HLEHpF H B HMHMHLEHH@0B xUHMv2 HuHuH H!7H Ht	HMK2 HHMH3
 H$   HP_^]돋
 @USWHHPHZ H3HEHHHUHU  HH\ He HLEHF H <B HEHEH} t	HM1 HUH   HE H# Ht	HM1 HHMH3

 HP_[]@SH@HZ H3HD$0LHHT$   Hd$( HHt"HLD$(HF H B HD$(HD$(HT$(H>   H|$( t
HL$(0 H|$  t
HL$ 0 HL$0H3Y

 H@[H\$Ht$WH 3HHH9tuH   HŞu9H  H貞u&H  H蟞uH8  H茞tH\$0Ht$8H _HP  Heu2H\$UHHPHZ H3HE3HIhH. H!]HLEHF H cB HEHEHEHEHUHMr  HUHMp  H} t	HM/ HEHtH8HM/ H} t	HM/ HHMH3
 H\$hHP]H\$UVWHH`HZ H3HEAHHUЃe H"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
