// @protobufjs/float/index.js
.com",
        "7luxeboutique.com",
        "7onetees.com",
        "7ravensglow.com",
        "7seventh.com",
        "7soulsdeep.com",
        "7strandsrawextensions.net",
        "7-strong.com",
        "7thandbond.com",
        "7thavenue.co",
        "7thirty.online",
        "7twenty5apparel.com",
        "7uiceshop.com",
        "7vncandleco.com",
        "7x7experience.com",
        "8.health",
        "8000kicks.com",
        "804escentialsllc.com",
        "805officials.com",
        "806girlboutique.com",
        "807ave.com",
        "808allday.com",
        "808kartel.com",
        "808rawpetfood.com",
        "80acremarket.com",
        "80eighty.com",
        "80sshades.com",
        "80stees.com",
        "80x20market.com",
        "815craftsupply.com",
        "818aquatics.com",
        "818skateshop.com",
        "81twentythree.com",
        "822decor.com",
        "828boutique.com",
        "828candles.com",
        "838apparel.com",
        "845motorsports.com",
        "84gem.com",
        "85apparelco.com",
        "8661hairboutique.com",
        "86and90.com",
        "86baits.com",
        "86squad.com",
        "87soueltique.com",
        "88acres.com",
        "88cardhouse.com",
        "88cosmetics.com",
        "88gear.com",
        "88musicstore.com",
        "88nightmarket.com",
        "88thandaugust.com",
        "88threadsnyc.com",
        "898westernwear.com",
        "89losoriginals.com",
        "89thandmadison.com",
        "89treasures.com",
        "8and9.com",
        "8apart.com",
        "8bitaesthetics.com",
        "8bitbeardco.com",
        "8dio.com",
        "8fifteen.us",
        "8fiftydesigns.com",
        "8figurewaistcouture.com",
        "8folddesigns.com",
        "8greens.com",
        "8otherreasons.com",
        "8sheep.com",
        "8thirteen.shop",
        "8twenty3boutique.com",
        "8voatico.com",
        "901honeysboutique.com",
        "903threads.com",
        "904badassmoms.com",
        "90x.co",
        "911healthshop.com",
        "912minds.com",
        "918boutique.com",
        "918hilltopdaisies.com",
        "922candles.com",
        "923collections.net",
        "925fit.com",
        "93brand.com",
        "93customs.com",
        "93playstreet.com",
        "941sportscards.com",
        "95clouds.co",
        "95northboutique.store",
        "97designco.com",
        "999club.com",
        "99centsubs.com",
        "99dollardress.com",
        "99jersey.com",
        "99wayzapparel.com",
        "9efer.com",
        "9five.com",
        "9greyhorses.com",
        "9lilas.com",
        "9livesracing.com",
        "9oclockteeparty.com",
        "9thavenueboutique.com",
        "9thcason.com",
        "9thstvintage.com",
        "a1bskincare.com",
        "a1fashionboutiquee.shop",
        "a-1locksmith.com",
        "a1poolparts.com",
        "a1ureboutique.com",
        "a22didit.com",
        "a24films.com",
        "a2bmotorsport.net",
        "a2ndmarriage.com",
        "a2zflowers.com",
        "a2zwholesaleapparel.com",
        "a3kbdg.com",
        "a3performance.com",
        "a4c.com",
        "a7.co",
        "a7xworld.com",
        "a90shop.com",
        "a91boutique.com",
        "aaadistributor.com",
        "aaasports56.com",
        "aaatoysandcollectibles.com",
        "aabowbabe.com",
        "aacaclothing.com",
        "aaconceptsco.com",
        "aacorvette.com",
        "aacreslosangeles.com",
        "aadancecostumes.com",
        "aadperformance.com",
        "aadreamboutique.com",
        "aaileenboutique.com",
        "aairah.com",
        "aajmolds.com",
        "aaketubah.com",
        "aakofii.com",
        "aalittlecreations.com",
        "aaliyahallurecouture.com",
        "aaliyahashanti.com",
        "aaliyahzcloset.com",
        "aallyn.com",
        "aamarirose.com",
        "aamiyahlashayy.com",
        "aamourcreations.com",
        "aandasoftpretzels.com",
        "aandatactical.com",
        "aande.shop",
        "a-and-h-aromatherapy.com",
        "aandkdesignsco.com",
        "aandlfluffyco.com",
        "aandmboutique.com",
        "aandmglittercreations.com",
        "aapc.com",
        "aaperfektatracktion.com",
        "aaprintsupplyco.com",
        "aarasessentials.com",
        "aarinandco.com",
        "aarke.us",
        "aarke-usa.com",
        "aaronasunsolo.com",
        "aarondavoe.com",
        "aaronkai.com",
        "aaron-michael.com",
        "aaronyapaints.com",
        "aarzdesigns.com",
        "aarzucollections.com",
        "aashopusa.com",
        "aastherapy.com",
        "aatbdesigns.com",
        "aatbow.com",
        "aatraining.fit",
        "aavrani.com",
        "aawcreation.com",
        "aawol.com",
        "abacaxi-nyc.com",
        "abacheclothing.com",
        "abaciorganic.com",
        "abacopolarized.com",
        "abacosmeticsupply.com",
        "abacusrow.com",
        "abaddezign.com",
        "abagailjoyceco.shop",
        "abandjjewelry.com",
        "abandonedcakes.com",
        "abarcainc.com",
        "abasiconcepts.com",
        "abasource.com",
        "abathhouse.com",
        "abathinflowers.com",
        "abattiste.com",
        "abayatmuna.com",
        "abbascloset.com",
        "abbas-love.com",
        "abbeybiketools.com",
        "abbey-glass.com",
        "abbieandharry.com",
        "abbiokitchen.com",
        "abbistreasuresboutique.com",
        "abboocandleco.com",
        "abbottnyc.com",
        "abbyalley.com",
        "abbyandjayboutique.com",
        "abbyapples.co",
        "abbyjodesign.com",
        "abbyjoycandles.com",
        "abbykahloworks.com",
        "abbykatehome.com",
        "abbyleighton.com",
        "abbyraeboutique.com",
        "abbyrarig.com",
        "abbyredaccessories.com",
        "abbysapothecaryco.com",
        "abbysbetter.com",
        "abbysclassics.com",
        "abbyson.com",
        "abbythehairstylist.",
        "abcamco.com",
        "abcbeadsupply.com",
        "abcfashion.net",
        "abcglassware.com",
        "abckidswear.com",
        "abcocrafts.com",
        "abcreations05.com",
        "abdesignsbyholly.com",
        "abdesignsco.com",
        "abdlr.com",
        "abdoer360.com",
        "abeair.com",
        "abearco.shop",
        "abeautifulyouboutique.net",
        "abebooks.com",
        "abeccagrace.com",
        "abehomegoods.com",
        "abeillebliss.com",
        "abejareinaboutique.com",
        "abeka.com",
        "abekahs.com",
        "abelastory.com",
        "abelpaulgeorge.com",
        "abelsworkshop.net",
        "abeotravelersclub.com",
        "abercrombie.com",
        "abercrombiekids.com",
        "aberlite.com",
        "abesofmaine.com",
        "abetterbuzzbrandgoods.com",
        "abeutifulyoufitness.com",
        "abeyoutifultreasure.com",
        "abglashwear.com",
        "abgresearchclub.com",
        "abhithenomad.com",
        "abibemade.com",
        "abidecandles.com",
        "abidesignstore.com",
        "abigailbellvintage.com",
        "abigailboutique.com",
        "abigailep.com",
        "abigailrosestore.com",
        "abigailsclipsandmore.com",
        "abigailscreationsboutique.com",
        "abigirlandco.com",
        "abigzails.com",
        "abiie.com",
        "abilenesky.com",
        "abillicosmetics.com",
        "abingdonco.com",
        "abirdandabear.com",
        "abirdtoy.com",
        "abkdco.com",
        "ablacklamour.com",
        "ablackmotherslove.com",
        "ablackstar.co",
        "ablanknote.com",
        "ableclothing.com",
        "ableground.com",
        "ableheart.co",
        "ablessedblinggirl.com",
        "ablessedglow.com",
        "ablissclo.com",
        "ablissfullybeautifulboutique.com",
        "ablossomfboutique.com",
        "ablyapparel.com",
        "abmat.com",
        "abmsshop.com",
        "abmxc.com",
        "abndntreality.com",
        "abnormalfresh.com",
        "a-bod-apparel.com",
        "abominablepedals.com",
        "abominabletoys.com",
        "abonicosmetics.com",
        "aboutbabyshop.com",
        "aboutface.com",
        "aboutthatlifeapparel.com",
        "aboutyoursmile.net",
        "aboveallluxe.con writeDouble_ieee754(writeUint, off0, off1, val, buf, pos) {
            var sign = val < 0 ? 1 : 0;
            if (sign)
                val = -val;
            if (val === 0) {
                writeUint(0, buf, pos + off0);
                writeUint(1 / val > 0 ? /* positive */ 0 : /* negative 0 */ 2147483648, buf, pos + off1);
            } else if (isNaN(val)) {
                writeUint(0, buf, pos + off0);
                writeUint(2146959360, buf, pos + off1);
            } else if (val > 1.7976931348623157e+308) { // +-Infinity
                writeUint(0, buf, pos + off0);
                writeUint((sign << 31 | 2146435072) >>> 0, buf, pos + off1);
            } else {
                var mantissa;
                if (val < 2.2250738585072014e-308) { // denormal
                    mantissa = val / 5e-324;
                    writeUint(mantissa >>> 0, buf, pos + off0);
                    writeUint((sign << 31 | mantissa / 4294967296) >>> 0, buf, pos + off1);
                } else {
                    var exponent = Math.floor(Math.log(val) / Math.LN2);
                    if (exponent === 1024)
                        exponent = 1023;
                    mantissa = val * Math.pow(2, -exponent);
                    writeUint(mantissa * 4503599627370496 >>> 0, buf, pos + off0);
                    writeUint((sign << 31 | exponent + 1023 << 20 | mantissa * 1048576 & 1048575) >>> 0, buf, pos + off1);
                }
            }
        }

        exports.writeDoubleLE = writeDouble_ieee754.bind(null, writeUintLE, 0, 4);
        exports.writeDoubleBE = writeDouble_ieee754.bind(null, writeUintBE, 4, 0);

        function readDouble_ieee754(readUint, off0, off1, buf, pos) {
            var lo = readUint(buf, pos + off0),
                hi = readUint(buf, pos + off1);
            var sign = (hi >> 31) * 2 + 1,
                exponent = hi >>> 20 & 2047,
                mantissa = 4294967296 * (hi & 1048575) + lo;
            return exponent === 2047
                ? mantissa
                ? NaN
                : sign * Infinity
                : exponent === 0 // denormal
                ? sign * 5e-324 * mantissa
                : sign * Math.pow(2, exponent - 1075) * (mantissa + 4503599627370496);
        }

        exports.readDoubleLE = readDouble_ieee754.bind(null, readUintLE, 0, 4);
        exports.readDoubleBE = readDouble_ieee754.bind(null, readUintBE, 4, 0);

    })();

    return exports;
}

// uint helpers

function writeUintLE(val, buf, pos) {
    buf[pos    ] =  val        & 255;
    buf[pos + 1] =  val >>> 8  & 255;
    buf[pos + 2] =  val >>> 16 & 255;
    buf[pos + 3] =  val >>> 24;
}

function writeUintBE(val, buf, pos) {
    buf[pos    ] =  val >>> 24;
    buf[pos + 1] =  val >>> 16 & 255;
    buf[pos + 2] =  val >>> 8  & 255;
    buf[pos + 3] =  val        & 255;
}

function readUintLE(buf, pos) {
    return (buf[pos    ]
          | buf[pos + 1] << 8
          | buf[pos + 2] << 16
          | buf[pos + 3] << 24) >>> 0;
}

function readUintBE(buf, pos) {
    return (buf[pos    ] << 24
          | buf[pos + 1] << 16
          | buf[pos + 2] << 8
          | buf[pos + 3]) >>> 0;
}
