// @grpc/grpc-js/build/src/load-balancer-outlier-detection.js
"use strict";
/*
 * Copyright 2022 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.setup = exports.OutlierDetectionLoadBalancer = exports.OutlierDetectionLoadBalancingConfig = void 0;
const connectivity_state_1 = require("./connectivity-state");
const constants_1 = require("./constants");
const duration_1 = require("./duration");
const experimental_1 = require("./experimental");
const load_balancer_1 = require("./load-balancer");
const load_balancer_child_handler_1 = require("./load-balancer-child-handler");
const picker_1 = require("./picker");
const subchannel_address_1 = require("./subchannel-address");
const subchannel_interface_1 = require("./subchannel-interface");
const logging = require("./logging");
const TRACER_NAME = 'outlier_detection';
function trace(text) {
    logging.trace(constants_1.LogVerbosity.DEBUG, TRACER_NAME, text);
}
const TYPE_NAME = 'outlier_detection';
const OUTLIER_DETECTION_ENABLED = ((_a = process.env.GRPC_EXPERIMENTAL_ENABLE_OUTLIER_DETECTION) !== null && _a !== void 0 ? _a : 'true') === 'true';
const defaultSuccessRateEjectionConfig = {
    stdev_factor: 1900,
    enforcement_percentage: 100,
    minimum_hosts: 5,
    request_volume: 100
};
const defaultFailurePercentageEjectionConfig = {
    threshold: 85,
    enforcement_percentage: 100,
    minimum_hosts: 5,
    request_volume: 50
};
function validateFieldType(obj, fieldName, expectedType, objectName) {
    if (fieldName in obj && typeof obj[fieldName] !== expectedType) {
        const fullFieldName = objectName ? `${objectName}.${fieldName}` : fieldName;
        throw new Error(`outlier detection config ${fullFieldName} parse error: expected ${expectedType}, got ${typeof obj[fieldName]}`);
    }
}
function validatePositiveDuration(obj, fieldName, objectName) {
    const fullFieldName = objectName ? `${objectName}.${fieldName}` : fieldName;
    if (fieldName in obj) {
        if (!(0, duration_1.isDuration)(obj[fieldName])) {
            throw new Error(`outlier detection config ${fullFieldName} parse error: expected Duration, got ${typeof obj[fieldName]}`);
        }
        if (!(obj[fieldName].seconds >= 0 && obj[fieldName].seconds <= 315576000000 && obj[fieldName].nanos >= 0 && obj[fieldName].nanos <= 999999999)) {
            throw new Error(`outlier detection config ${fullFieldName} parse error: values out of range for non-negative Duaration`);
        }
    }
}
function validatePercentage(obj, fieldName, objectName) {
    const fullFieldName = objectName ? `${objectName}.${fieldName}` : fieldName;
    validateFieldType(obj, fieldName, 'number', objectName);
    if (fieldName in obj && !(obj[fieldName] >= 0 && obj[fieldName] <= 100)) {
        throw new Error(`outlier detection config ${fullFieldName} parse error: value out of range for percentage (0-100)`);
    }
}
class OutlierDetectionLoadBalancingConfig {
    constructor(intervalMs, baseEjectionTimeMs, maxEjectionTimeMs, maxEjectionPercent, successRateEjection, failurePercentageEjection, childPolicy) {
        this.childPolicy = childPolicy;
        if (childPolicy.length > 0 && childPolicy[0].getLoadBalancerName() === 'pick_first') {
            throw new Error('outlier_detection LB policy cannot have a pick_first child policy');
        }
        this.intervalMs = intervalMs !== null && intervalMs !== void 0 ? intervalMs : 10000;
        this.baseEjectionTimeMs = baseEjectionTimeMs !== null && baseEjectionTimeMs !== void 0 ? baseEjectionTimeMs : 30000;
        this.maxEjectionTimeMs = maxEjectionTimeMs !== null && maxEjectionTimeMs !== void 0 ? maxEjectionTimeMs : 300000;
        this.maxEjectionPercent = maxEjectionPercent !== null && maxEjectionPercent !== void 0 ? maxEjectionPercent : 10;
        this.successRateEjection = successRateEjection ? Object.assign(Object.assign({}, defaultSuccessRateEjectionConfig), successRateEjection) : null;
        this.failurePercentageEjection = failurePercentageEjection ? Object.assign(Object.assign({}, defaultFailurePercentageEjectionConfig), failurePercentageEjection) : null;
    }
    getLoadBalancerName() {
        return TYPE_NAME;
    }
    toJsonObject() {
        return {
            interval: (0, duration_1.msToDuration)(this.intervalMs),
            base_ejection_time: (0, duration_1.msToDuration)(this.baseEjectionTimeMs),
            max_ejection_time: (0, duration_1.msToDuration)(this.maxEjectionTimeMs),
            max_ejection_percent: this.maxEjectionPercent,
            success_rate_ejection: this.successRateEjection,
            failure_percentage_ejection: this.failurePercentageEjection,
            child_policy: this.childPolicy.map(policy => policy.toJsonObject())
        };
    }
    getIntervalMs() {
        return this.intervalMs;
    }
    getBaseEjectionTimeMs() {
        return this.baseEjectionTimeMs;
    }
    getMaxEjectionTimeMs() {
        return this.maxEjectionTimeMs;
    }
    getMaxEjectionPercent() {
        return this.maxEjectionPercent;
    }
    getSuccessRateEjectionConfig() {
        return this.successRateEjection;
    }
    getFailurePercentageEjectionConfig() {
        return this.failurePercentageEjection;
    }
    getChildPolicy() {
        return this.childPolicy;
    }
    copyWithChildPolicy(childPolicy) {
        return new OutlierDetectionLoadBalancingConfig(this.intervalMs, this.baseEjectionTimeMs, this.maxEjectionTimeMs, this.maxEjectionPercent, this.successRateEjection, this.failurePercentageEjection, childPolicy);
    }
    static createFromJson(obj) {
        var _a;
        validatePositiveDuration(obj, 'interval');
        validatePositiveDuration(obj, 'base_ejection_time');
        validatePositiveDuration(obj, 'max_ejection_time');
        validatePercentage(obj, 'max_ejection_percent');
        if ('success_rate_ejection' in obj) {
            if (typeof obj.success_rate_ejection !== 'object') {
                throw new Error('outlier detection config success_rate_ejection must be an object');
            }
            validateFieldType(obj.success_rate_ejection, 'stdev_factor', 'number', 'success_rate_ejection');
            validatePercentage(obj.success_rate_ejection, 'enforcement_percentage', 'success_rate_ejection');
            validateFieldType(obj.success_rate_ejection, 'minimum_hosts', 'number', 'success_rate_ejection');
            validateFieldType(obj.success_rate_ejection, 'request_volume', 'number', 'success_rate_ejection');
        }
        if ('failure_percentage_ejection' in obj) {
            if (typeof obj.failure_percentage_ejection !== 'object') {
                throw new Error('outlier detection config failure_percentage_ejection must be an object');
            }
            validatePercentage(obj.failure_percentage_ejection, 'threshold', 'failure_percentage_ejection');
            validatePercentage(obj.failure_percentage_ejection, 'enforcement_percentage', 'failure_percentage_ejection');
            validateFieldType(obj.failure_percentage_ejection, 'minimum_hosts', 'number', 'failure_percentage_ejection');
            validateFieldType(obj.failure_percentage_ejection, 'request_volume', 'number', 'failure_percentage_ejection');
        }
        return new OutlierDetectionLoadBalancingConfig(obj.interval ? (0, duration_1.durationToMs)(obj.interval) : null, obj.base_ejection_time ? (0, duration_1.durationToMs)(obj.base_ejection_time) : null, obj.max_ejection_time ? (0, duration_1.durationToMs)(obj.max_ejection_time) : null, (_a = obj.max_ejection_percent) !== null && _a !== void 0 ? _a : null, obj.success_rate_ejection, obj.failure_percentage_ejection, obj.child_policy.map(load_balancer_1.validateLoadBalancingConfig));
    }
}
exports.OutlierDetectionLoadBalancingConfig = OutlierDetectionLoadBalancingConfig;
class OutlierDetectionSubchannelWrapper extends subchannel_interface_1.BaseSubchannelWrapper {
    constructor(childSubchannel, mapEntry) {
        super(childSubchannel);
        this.mapEntry = mapEntry;
        this.stateListeners = [];
        this.ejected = false;
        this.refCount = 0;
        this.childSubchannelState = childSubchannel.getConnectivityState();
        childSubchannel.addConnectivityStateListener((subchannel, previousState, newState, keepaliveTime) => {
            this.childSubchannelState = newState;
            if (!this.ejected) {
                for (const listener of this.stateListeners) {
                    listener(this, previousState, newState, keepaliveTime);
                }
            }
        });
    }
    getConnectivityState() {
        if (this.ejected) {
            return connectivity_state_1.ConnectivityState.TRANSIENT_FAILURE;
        }
        else {
            return this.childSubchannelState;
        }
    }
    /**
     * Add a listener function to be called whenever the wrapper's
     * connectivity state changes.
     * @param listener
     */
    addConnectivityStateListener(listener) {
        this.stateListeners.push(listener);
    }
    /**
     * Remove a listener previously added with `addConnectivityStateListener`
     * @param listener A reference to a function previously passed to
     *     `addConnectivityStateListener`
     */
    removeConnectivityStateListener(listener) {
        const listenerIndex = this.stateListeners.indexOf(listener);
        if (listenerIndex > -1) {
            this.stateListeners.splice(listenerIndex, 1);
        }
    }
    ref() {
        this.child.ref();
        this.refCount += 1;
    }
    unref() {
        this.child.unref();
        this.refCount -= 1;
        if (this.refCount <= 0) {
            if (this.mapEntry) {
                const index = this.mapEntry.subchannelWrappers.indexOf(this);
                if (index >= 0) {
                    this.mapEntry.subchannelWrappers.splice(index, 1);
                }
            }
        }
    }
    eject() {
        this.ejected = true;
        for (const listener of this.stateListeners) {
            listener(this, this.childSubchannelState, connectivity_state_1.ConnectivityState.TRANSIENT_FAILURE, -1);
        }
    }
    uneject() {
        this.ejected = false;
        for (const listener of this.stateListeners) {
            listener(this, connectivity_state_1.ConnectivityState.TRANSIENT_FAILURE, this.childSubchannelState, -1);
        }
    }
    getMapEntry() {
        return this.mapEntry;
    }
    getWrappedSubchannel() {
        return this.child;
    }
}
function createEmptyBucket() {
    return {
        success: 0,
        failure: 0
    };
}
class CallCounter {
    constructor() {
        this.activeBucket = createEmptyBucket();
        this.inactiveBucket = createEmptyBucket();
    }
    addSuccess() {
        this.activeBucket.success += 1;
    }
    addFailure() {
        this.activeBucket.failure += 1;
    }
    switchBuckets() {
        this.inactiveBucket = this.activeBucket;
        this.activeBucket = createEmptyBucket();
    }
    getLastSuccesses() {
        return this.inactiveBucket.success;
    }
    getLastFailures() {
        return this.inactiveBucket.failure;
    }
}
class OutlierDetectionPicker {
    constructor(wrappedPicker, countCalls) {
        this.wrappedPicker = wrappedPicker;
        this.countCalls = countCalls;
    }
    pick(pickArgs) {
        const wrappedPick = this.wrappedPicker.pick(pickArgs);
        if (wrappedPick.pickResultType === picker_1.PickResultType.COMPLETE) {
            const subchannelWrapper = wrappedPick.subchannel;
            const mapEntry = subchannelWrapper.getMapEntry();
            if (mapEntry) {
                let onCallEnded = wrappedPick.onCallEnded;
                if (this.countCalls) {
                    onCallEnded = statusCode => {
            
Q,hrVL9US?Fc[ʔԈ5|q|z)K*:ޓ>D@gɷtt\a1(5M(าl	b.bF7/iMW`갌rRO6VU%x)lht ƄǬHy$ptt\C:<~P[
ÂJaV USS>Ձ 
lyhԉSj^tvjխZqE

Ǯ18Mt9'%m0,M&T}H/d=1PMEnW~ڴ2$PRSC޿yKCg62֣g:r%W_תo?:zZF8B}$(ƙV[?
ƃ.
M
.n..:K	䀶QEߒv!]IgҦ`$N;,5D
"ZSȸК߻=+,kDXsrSźZB015de-n=U^I"v_\hC5I{@>wUR*s*M /P5	`.Z޽NE^sؾ0=&GHIDTDQd\	=T>+O/WJ4%ܴjHzҠکOGE"q
>AxYQcL,tba32r]tBGP[1z/b{] #>M"81ޮk8$pxGRi#	p*b={8 +V:Y֜J@/QoC
a:v-fy#\"T**iDhJ\X2(
=Cf9	ć>5Zl7y۶ͩAyY|;<Mmݱ7<B8C2vk "0Vq.Do ȐK%=ސwp ZZ8Kֽ+5Eau-FI}mxrb.M/Ep~`QMU'OY(=F	8UU.'	~(._H} l*}DѾB;Kh
m7Mqٔ  ]i[ʒ_a\"]1d9	Bpn5!hx=}nI-/s?LrNkuU[Uvyĳ #	';8	<Kb!q*+]Mqea Bu.H6ҍ*z-'햖HLQ[t4{Yk@MPpҥFnԀieSMIhe3`yζ/l[D↶Gu[ڌ&3xDmg\oF _RVJq2,f9 \Q^CAArmfCq;̃l-3[цM;렮@r(`mLE.&ѓNF0/
x8`~l wY3~
|c-sHAlg|fmviΘckH+w0@ifl:fǤEXOO'Z[kTzlX;T'>a)(.C%]~;A
2;pk2=.j7o#+M1ܦX?-_i|ܦ=ucSt.v`~OMv*'R Ĕy xڸ*TP^S{@r~uqPsC#6[cT`1˙d(BߨBS*rDJ1ϛ#*|<^@fJ0]:N5Yi'y,eJtDQT2Zy8	j&iر%XN,X`u+KZFbJ
'W#BZS
&E|!iV#bdO!lbSb?p@0́3fPrz.$U3'SfnX3Rͷ+LN.q*7j@0`6"ppN=5yn
0Δ	Z3`_*>q2Yrۄ
נfEb"h}zMN*(3c\%zob,lt.kD,.F~SEXZ1(pN޾p3P]* jC6FRߴPj<omwq7E/lȞM#=&!6k3A4yTk i0R'T1d5%͏5_36ە^	SM&Cf3]Ú=$tIE-ĂirhSGo@L-fX$	wR0$W!Z$
צֺ/QXUIrxBƢ&Nٓ̅K	T
X	8<TG
FKPb˦-O8@?0ӃBY@="=5R]\+<YnvO;?7哟^tk?*㿎hPKwkw{{vs|qst~<9z\.ol{C8?>FhzS]qwo~.}KAdk#1_ͯaJǧ^+]yfTT=;;}5֟ߣ_;_?Z7߽ޢ<=o\5w]9|
4%{$	(ؐk0
sp=}S3~ZjboJ_H2ײ0
A_0Md5ƀ'n'1u@qmnڟs7G`DӻLl-ϩ'5zN<_*ú~bR[$5C$aL;[ |ӹ?Y8ӣa6T`E9'%Fbg	퀄fRs˚@AjԄ%?f\bp$~ϼS	ȥ	)IL^Q;0y8Yg 33¼y11Q̀sUwRYϼj%tON:';mÜϞTZ&w]M:	
,ȣ)S0P_PTpFp
 @dIB50R3'+OzV	4Ϙ,'I<7{qYdp(?e^Ȍ2z"X NHnyg72:W5.qV5U.c}3xb73f 90byuvbӗ@Bl>n"rli;w#s_gUB
eIV,6lTqGKL9^bhEEORgk(S<Ͱʂ?C}_RpIVԾZ;(;p?QO*8:*WsռR9ቇ7
iDSڽ:oE-ewZ=,q1SNm>1
ZCcjQ:da	jф9,W0.Ǐ
J(	EClSM5|5Ի3kO(k;0sYKߡ0Y)t;N~yNh7"Gԟ j,m0v"a$n
D@hBfsw8h1ď9u?昵t͌Q_p'd^ۡ͠Q+X#ʣ}  
{_r<!Dh4:/8>?ol{ơXz~OJ@&HJ
/
Mŏhs+QwNbNMSYóc;n+H6p^+ wߗ+1>~?ʞjԏ%N>tUmmh귛`l[kϹ7n;x;k{R5D_Иٰ&Y{}YْLyL\-{Lzq*21uY[Ϻ6"t-_|[N@: .|˺#h=At:߆⒯'^ZMO4ړ8-mo3tv>:hP'&3,bM4i4Sj&ke4c'G]-h
Ĵbc]-g̗%Ž#e..k
1&D{]a1٪\8ܩFqH+19L-$P1rڶ80MMqrS bJކkƓI7yǗ`GEKMPiehXQEVRDHl S`/i6V1c#L	ۥ
Ŀ&Ik\ZՓUlR]i]cV[ɺ6d]
<]pMن6c`Le\AfVh1΄l[+`hC kGY1&)&eZkGM<_m^2pqxȓ10`᫔.T+_Dmcbv}[e88:WYq:ķxbQt5T\Y"I_iEfOk;N'U3&њ0=jRǘ oCX/"#a_k/Aw%Sa[X3|^G.屏>rlg	4^V p}GeWÒo@McSٓ un
jH+ؕ|=7.d8Et|b<a{vV	gOWՈD<ؠrإzJ4%
\:-Q4P;Nm;ڴlZij4~s@e:Q[PJ	~ő;'\0/Ȍ&>a{ۚt-ٛ5ŗ$fem4tU$a4NbǊPK"fng'=WŲgW}n|S{WScݑCksq{JZ 8GN]٨{
ZimPM4N/rU]溃>n
/=`H,YmȦ倪jʪͶgDrolz`혆Xb{
r,PwۈY9xn>ȖWVȄ[\+)q,fOL;eUBFYF}  <,H/;G"A]xHc3df&@$Zn+Q@D?EZ4/ʺFF~>&- zp(1!?vRDgb{S,S0g[gGGŰ' ypI1242F,70^6zdΣ[h0Y"3J)94ĉŪ4y94 y Hqe;K[#w|-1Bޣ&qE=wK; %D╳$6]bQm?cQs::^/D1v^;Uͨ
X)}^K> \j$JeE+}Zi6r["!њ|鰔~>W*Cj!)\z"O`􋟇n%)D5c=;ў݉ 
"["Fk()Ү!Z5KQnpbfa,?Y6l.'|̾xבּuO^jFdY$ϾK$mmq$,őLox\8"3՗; VU}aަ{Ֆe]eI4}ET"M洁/l,xsڈ?uC;8{amdӌo!S"wTG`#̫8u
muIh>v.Pn}}n.>h]"P,qԥzV_[,u뵇ca\[~5֮t4\ZJ0kyvePan*L%yK~ۇ?=׿xH}X\uɑiDQH$(/nUP@k'ţ0;-Y9h'_LƹG$'T0Sy2T?#tܟSԦiikb޼gWw#?gҍє57{cȏd(tkl%cͥo~ON{e95D(|Lg8f.؁xr&0]i&˻=|ɘN/=q]]L哝t-/@
(~¤a5S?7>54F&]75az֍~M8\mxpee󨹲esj?_Kx].~Npc3NlB_Oh4J
?x,p)F	&_ÕmqIsa0m7ϯN;Uޘ1EL-H͠sǩst7S/qq(Ȅ;
M(Z#eT1~ꦤj;aU0i^CbMSuU]M0Mf8}182F:O?%wSfF+Ax(g-v+w=N݂yk쵹""";,՟i,{͂ӻ+AQ,Mw}GLU;CsdN[~3QՌM\]0%>GYr`?DVx?i:N_^sotc^6cVk$l
`i;ܕuA(Lƴy}56J8VbcjB f7E6mq嶦;TqϧbWŶ`<wK4	vOyqĢkb۴ONobG(+Mm"S?
OWZ(&ʘɖ
	]VVSOS+*Ԥq+ǚLrՂ~)Ff?+WsW޼*^QICk~s"¾Kƃ$/'[(C6q, un^n;V#=.4q=AԿZA%YK`AkđP7AB(>hoۣF
Ǒ憞.nvG7mHq(ik5
H<S(2Շgfn{Q  ]iw_uEcc;jlG KxI~kVK`'3zoUWwW?eSǨ1GI<*$ýpurxO|"4J'cXջHaGQg8iR&/˱׎yW0)^^_L:M*OMTOXi p2=$mL
og?mO%_2I>/	=/xdNDm4$M벓.؉OMz~/ӸC'El#U9Y7c1ݗ,mNSFW^ UUZ0SKV%Qs
]79¼4½$sʪ'1>N,&z	uL>zSB9YМ}|G|w˅q'ɳnqFOYЖ}tC9sf%[Vx2
Ս*2F8猭o
Bo8	Q'*!M+_YFr]֊\>0'Kghͧ/){4-X/ne-
h9MYǣ5|PB}ZF:	ɀCI'!76ᴓVKڬ|CG"B稞E%:3<`u,pih|k/=٭w-y F!=%R,آQ}Y^T&4[Оn>)IkN~tkYaR퍧w\R893gI"w&<r9PL#ұ8>iFM32ag,@@x. 0	(,rw?AU6٦AQlk-a֎r'Qqd<nC_ڼKa<nXO	^߇0<Sff0S7{NGR{Aci";MGב_Bn0r)ہR7ק_mgY;l'!6~C̪|Gc9+1X
,eqL+C'xZ`	Hn%p}mV;d	`/GQXZ
M6'聶0/XT/~̐z;,X d$X43Ud|9,^B4e-ιN"a艎'mT䩵M[W j5OzYű<Pmy(g}"Z)t):i,qZ +:YI2b009}\L1߈³kk%hzH`Z)4&&cM7~=w"zFŃqbڎ*6aRaTv7{vehv(l;5ϱ.u]~#9jk]8 pWA~33Xo\ /*^@D~$<C`y.'Q&NavHW#S4:yd<>Gh<:0QD%R Fp2(f'P;޲Yd_'Mglˇ49H#ށWMw<^\YR"͌f[/fHmKMAӻL^56Gdb]F7Celq@gvnq_c-3 7}s#4J7gZ@E/.OߝחO^_6XΊQ[LћU6E2/Z Jݸ@^Ά(|@?N]ҡ%d KӺ?2A6L9D"vM(Fƪ\Mx7fLTЖGDГ7o^<Ab8و&G/Ydb2O鞕^s
 uXx6v@+'}cTknT2N91[y:&^g^02BapH}슒SeC/C4rR`|2}ܱY#

Hi!WWgG̩s<u`#A!#ex9r֥(#4"(tsE[s:Ih4a&}sG=tx-wմ
HI5^*ZA֟8D0g~\r%#!+N>6d|PGzlάp/	H?00l%m:hƈam'M{^xܛ
]n2,E+?X8Sk #;!r->dwWppa`-
*"ډ7;iI	fJ'URRt	"
,vw%O=v;V(N=ǱV`P̿hD_6"l*jSJXIƃ3rí.RjT%$'KM&}Jן^3P~(uQݭ,SQFQdxʪ֦ky| zDnKaC/Ah>N6c{zAbN@vtΫJzB_"	W#G7*mL'!b:.0KBLN~u׾[:yIe+4\2re;lxP:ѲoO=ȃ@!QlĠ(٫MfvaH힭Ȝy,;Nt~,lB^\|UBˉMV}1鵠0{hv^\:JLg\.[Hj!=yZfP%kZVGN.료m`Hٴa?4n)Zݬ껛Nޯmo
fV߯jNeXښ\<u8@8!GP<)_;N}g{kks{wmnnn"B.|ڎSoUK Im٬lWg EuB~U٭^Wwk۵^+#uնvZwJΞ?mB(2
"lw5{-dM_6}ٗv+,T%WPO>x5HDøv_榣]\oH2/\ٶ8֞S6a08h6{ћoa3?OLM8(
^+JLw/
':^EBX,KyMˑd>L{ϑZӠ5t'u%y_ ͏y}o,@>Ƒ~rD ӎeimDgr"Fѥ"D5FUV4:qbBsͤre@s(߲<ыf0`5v pbթ"h'0L9O3~
M>IT%B#ݕUr|Qv>)r^Ds77e!p&]0*0?U
=h2v^ǛQrKP"]L)QA2ٯϪruQWB]X+'.hM[QnabQ+i4Oĸ.ͳSԭ>GBR348fŶ0.TQu_Hf?Vﺖyd@\
!8W1}kDw,UZfp\VkT
O`L~V+Y v[W^n7*`+C[xeчer
 wF(q2iFlЋ6/W䥝+M >0F5TȀy"A0lbI!7XY:cuȻ[,v*4+HBacz)>K Xk5?Ȇbb%q \=?^@fLhu6bP37YO7|{a71!Rrd%_KYTC\<
PG}6 9Q>j
cN@, R*!I,,oTE9F/3ݡɪtxQ[舐$1Q?0GbsA5{S2EUBMáOB .&!izuaVu:
# 	2bˉ2ͳ3@k~@Sr@Yl=\1T,b(eقLf>S]l`xeN"LQ/:F;n	JwQEmQj{4gŕg.v[;HߨҦ>!S7Ԋ7q%"bh8	So9:S[syn<-4B}BdL2"0er);00^;;:	
uC:oZqD}>Wxj[=YP[@͹"̋
a!+D ቓl;))qp"Ez.DMdrcWUcΞ&3hY]V.Ns
wSb$ܛ`s<ҡg/` s]]5~izا'4^	
;*>jW%%%	"xC8P#&]kځ ~? AEe0Mw*wo
dzmġm2=o81</4@j{M@o'8tkCuߪ+_S(XSHe8:geEU*~f-C$3͙>^t,b,g`j1a崸穹`bݿ+ȥAэ/Kz:rA`<#V/LXR7oh""3N%,_B1 av3fS,$3e߱4=o#ft*CXs'c9}Ч{At1Ba\
qQK»6.;ihhg|β%\
/rRG;8L@T.tkD}L4
1omBטЏ{RxK6d@h|k!85R 35.ƈނ+Tq|qݮ!p!Oq)QJ*Rt%hL}CANa	ڸ:y`+\L<En@22^YS5++݆!4[& zA^yuCEq! TC>D#e	Oxl9r~-]DNy/j+٩moor~CPAby~Z|h]v'C+(wTydJUG4Iדăy}(
aqkrt" 	Bb+lǼ߂2}>?
..Jy$(4B';1<Hq7vpio}vbAZCa
 ,	e!T^/(YFmMB?p.<G_T0Pa
ZLgHxE&!1>:˸L.ΐZl)fBB8$ .Lg#n`
dh6ڞ.U(s4~B	h34?F)^iQ*L(@Q$B^ʢDΨIݞ%ww../JZˎ`}#ۥ>B-R "56lƝ[/Mum)Y-b.TJ$Y	9
?0C-']5T76s  ]S:4m< ;;N.6Znxaly(}COvwv;SbKÒ,}B<4a	v5tޓ)ȤhVX>Y;qHLIすhao%BB~QH_0Tlp%*K~}yCwIzWhfYz<S,Jd,FnQ}Y{8
(&ǩc?T&Ur|!E9
/jϣd|^!<aY`e4Lqcd忚k
