// firebase-admin/lib/remote-config/remote-config-api-client-internal.js
/*! firebase-admin v11.11.1 */
"use strict";
/*!
 * Copyright 2020 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseRemoteConfigError = exports.RemoteConfigApiClient = void 0;
const api_request_1 = require("../utils/api-request");
const error_1 = require("../utils/error");
const utils = require("../utils/index");
const validator = require("../utils/validator");
const deep_copy_1 = require("../utils/deep-copy");
// Remote Config backend constants
const FIREBASE_REMOTE_CONFIG_V1_API = 'https://firebaseremoteconfig.googleapis.com/v1';
const FIREBASE_REMOTE_CONFIG_HEADERS = {
    'X-Firebase-Client': `fire-admin-node/${utils.getSdkVersion()}`,
    // There is a known issue in which the ETag is not properly returned in cases where the request
    // does not specify a compression type. Currently, it is required to include the header
    // `Accept-Encoding: gzip` or equivalent in all requests.
    // https://firebase.google.com/docs/remote-config/use-config-rest#etag_usage_and_forced_updates
    'Accept-Encoding': 'gzip',
};
/**
 * Class that facilitates sending requests to the Firebase Remote Config backend API.
 *
 * @internal
 */
class RemoteConfigApiClient {
    constructor(app) {
        this.app = app;
        if (!validator.isNonNullObject(app) || !('options' in app)) {
            throw new FirebaseRemoteConfigError('invalid-argument', 'First argument passed to admin.remoteConfig() must be a valid Firebase app instance.');
        }
        this.httpClient = new api_request_1.AuthorizedHttpClient(app);
    }
    getTemplate() {
        return this.getUrl()
            .then((url) => {
            const request = {
                method: 'GET',
                url: `${url}/remoteConfig`,
                headers: FIREBASE_REMOTE_CONFIG_HEADERS
            };
            return this.httpClient.send(request);
        })
            .then((resp) => {
            return this.toRemoteConfigTemplate(resp);
        })
            .catch((err) => {
            throw this.toFirebaseError(err);
        });
    }
    getTemplateAtVersion(versionNumber) {
        const data = { versionNumber: this.validateVersionNumber(versionNumber) };
        return this.getUrl()
            .then((url) => {
            const request = {
                method: 'GET',
                url: `${url}/remoteConfig`,
                headers: FIREBASE_REMOTE_CONFIG_HEADERS,
                data
            };
            return this.httpClient.send(request);
        })
            .then((resp) => {
            return this.toRemoteConfigTemplate(resp);
        })
            .catch((err) => {
            throw this.toFirebaseError(err);
        });
    }
    validateTemplate(template) {
        template = this.validateInputRemoteConfigTemplate(template);
        return this.sendPutRequest(template, template.etag, true)
            .then((resp) => {
            // validating a template returns an etag with the suffix -0 means that your update
            // was successfully validated. We set the etag back to the original etag of the template
            // to allow future operations.
            this.validateEtag(resp.headers['etag']);
            return this.toRemoteConfigTemplate(resp, template.etag);
        })
            .catch((err) => {
            throw this.toFirebaseError(err);
        });
    }
    publishTemplate(template, options) {
        template = this.validateInputRemoteConfigTemplate(template);
        let ifMatch = template.etag;
        if (options && options.force === true) {
            // setting `If-Match: *` forces the Remote Config template to be updated
            // and circumvent the ETag, and the protection from that it provides.
            ifMatch = '*';
        }
        return this.sendPutRequest(template, ifMatch)
            .then((resp) => {
            return this.toRemoteConfigTemplate(resp);
        })
            .catch((err) => {
            throw this.toFirebaseError(err);
        });
    }
    rollback(versionNumber) {
        const data = { versionNumber: this.validateVersionNumber(versionNumber) };
        return this.getUrl()
            .then((url) => {
            const request = {
                method: 'POST',
                url: `${url}/remoteConfig:rollback`,
                headers: FIREBASE_REMOTE_CONFIG_HEADERS,
                data
            };
            return this.httpClient.send(request);
        })
            .then((resp) => {
            return this.toRemoteConfigTemplate(resp);
        })
            .catch((err) => {
            throw this.toFirebaseError(err);
        });
    }
    listVersions(options) {
        if (typeof options !== 'undefined') {
            options = this.validateListVersionsOptions(options);
        }
        return this.getUrl()
            .then((url) => {
            const request = {
                method: 'GET',
                url: `${url}/remoteConfig:listVersions`,
                headers: FIREBASE_REMOTE_CONFIG_HEADERS,
                data: options
            };
            return this.httpClient.send(request);
        })
            .then((resp) => {
            return resp.data;
        })
            .catch((err) => {
            throw this.toFirebaseError(err);
        });
    }
    sendPutRequest(template, etag, validateOnly) {
        let path = 'remoteConfig';
        if (validateOnly) {
            path += '?validate_only=true';
        }
        return this.getUrl()
            .then((url) => {
            const request = {
                method: 'PUT',
                url: `${url}/${path}`,
                headers: { ...FIREBASE_REMOTE_CONFIG_HEADERS, 'If-Match': etag },
                data: {
                    conditions: template.conditions,
                    parameters: template.parameters,
                    parameterGroups: template.parameterGroups,
                    version: template.version,
                }
            };
            return this.httpClient.send(request);
        });
    }
    getUrl() {
        return this.getProjectIdPrefix()
            .then((projectIdPrefix) => {
            return `${FIREBASE_REMOTE_CONFIG_V1_API}/${projectIdPrefix}`;
        });
    }
    getProjectIdPrefix() {
        if (this.projectIdPrefix) {
            return Promise.resolve(this.projectIdPrefix);
        }
        return utils.findProjectId(this.app)
            .then((projectId) => {
            if (!validator.isNonEmptyString(projectId)) {
                throw new FirebaseRemoteConfigError('unknown-error', 'Failed to determine project ID. Initialize the SDK with service account credentials, or '
                    + 'set project ID as an app option. Alternatively, set the GOOGLE_CLOUD_PROJECT '
                    + 'environment variable.');
            }
            this.projectIdPrefix = `projects/${projectId}`;
            return this.projectIdPrefix;
        });
    }
    toFirebaseError(err) {
        if (err instanceof error_1.PrefixedFirebaseError) {
            return err;
        }
        const response = err.response;
        if (!response.isJson()) {
            return new FirebaseRemoteConfigError('unknown-error', `Unexpected response with status: ${response.status} and body: ${response.text}`);
        }
        const error = response.data.error || {};
        let code = 'unknown-error';
        if (error.status && error.status in ERROR_CODE_MAPPING) {
            code = ERROR_CODE_MAPPING[error.status];
        }
        const message = error.message || `Unknown server error: ${response.text}`;
        return new FirebaseRemoteConHSH;t(ILHH;tL9`tHHH;u@ulmHH;StNAE A8AMIHLBMtI@ HBHuH!B I`  Ha  HMHtHN HMHHJHuHd@H8I9HUIN(*3@t6@uHEHEEGDEWEHuHEHD$ HUIN`H]LHt:HHC HHw HWEHuE3LEI3IB(	 LHuHMH3ˈ H$   HĀ   A_A^A]A\_^]H\$HL$UVWATAUAVAWHHpIMLI8u{3uI0 څTMH
/3 I̾'  L=3    H;vH3 IH@0X	 @@HHEȉuL}MECAAF  H
S	 m	 HHR	 e	    I0 څTML=2 3HtR3IM 	 u(	 =  tMǋHM5   HlWEHuECWuH
42 IH;vH2 ='  IH@0k	 @HHEE='  L}EMK@{tH9{   93   3HM@HqHE   IM HMPLh0LmI0	 @}uH1 IԾ  H;vHv1 IH@0	 @@HHEuL}EEMMLEHUPHMI	 AKANIH$   HpA_A^A]A\_^]H(y t	H	I	 H(H	H%:	 H%	 H%	 HhHM0 H3HD$P   A4u~   A8urHI(E3E33+	 u\	 Lp0 HL$8>  D$HHQ	 HD$ HP	 HD$(D$0   D$4   tH|$Hu|$8 uHL$PH3^ HhLD$ HQ	 HL$8!    H\$UVWH HHپ   q0r0LHI(3E3l	 u<HK`N	 Hg  H   HtHx H   H   HK`sX$	 H\$PH _^]H(y t	H		 H(H%	 H(   	 u	 	3   H(H\$Ht$WH WHI@H0 څTMHIHvI IH@0D	 HHsHt$8HCHÉ;H\$0H _HyrH	HH\$ UVWATAUAVAWH0IHHHL$(E3L1LqLqLzM`IHO,<HBI+L;wTH;JwNJKLrHB   D2HH{rHIxrI0MD$J8HY Lk   I+L;wHA AHKMpI@   E0H3MD$J>H HrH?MHHHHI+I;   IHH;wH   H;HBHHHHBH   r!  
HtȂ LL3LkHkHrH?MHIz H~rH6MD$K>H_ HH$   H0A_A^A]A\_^]C   HN	 Åy%  =   u"ɋ     TP  tGt<6  -  D  #    (     ø(   3Ãt     t  t|  A
ø   ø  ;Y     {      CN   $    }     t[tt]s     ø   ÃP   >  )  t"   ?  t	Z   ø    ø   ø  ;]            t7   
t#  t.     ø'   ø   ø)   Á  t/   
ttt   '     øi   ø;'  ;      ('  ;atYa	  tKtF  t   t0t%tt#     ø   ø
   ø	   ø   ø   ø   Á3'  tOtDt9t.t#tt
   Azø   øs   øm   ø   øg   øp   ø   ø   øF'  ;ct[='  tMtBt7t,t!ttu`j   øu   øv   øt   øe   ød   øf   ø   øl   ÁG'  tDt9t.Dtttøn   ø&   øk   ø~   øq   øw   H(LA)u	DLJHb HH(@SVWH@II؋M   IuA  yHd$0 A   |$(D3H\$    	 HcȅuDLI	 HHV  6H~/|
t|
uH H~Hc|.uD HH@_^[H	H%	 @SUVWAVH HDIHL;   HA   HsAHYB ; kHHH;w   HH;HBHHHHBH   r
  HHt} 3AH7LH_HHoՑ 3 H A^_^][  H\$UVWATAUAVAWH0LaHHLL$ I+HH;  LiJ,"HHH;wIHHH+L;wJ)HH;HBHHHHBH   r
P  HHtA| 3H^M4<H\$ M<<LHn$   MHIrTHHސ LD$ Iې IUA H   rHKH'H+HCHw
HH{  	 H荐 LËI茐 A H>HH$   H0A_A^A]A\_^]]  H\$ UVWH@HiHHAHH+H;w2HyH*HAHrH1AH.L H7H) DD$ LDD$0>H\$xH@_^]H\$Ht$ UWATAVAWHl$H   H% H3HE'DHHME3LeLd$0Dd$(HEHD$ A   D3ҹ   	 uFELhF	 P&HM  L'LgLgHEHHD8$0uLHUH    H]H]Ld$8Ld$0Dd$(Ld$ HDL33	 LcuBELE	 V'HMs  L'LgLgHEHD8$0uLHUHt    LeLeLeM3HM)HEH}HCELd$8Ld$0D|$(HD$ DLE33	 D   ELPE	 P&HM  L'LgLgHEHD8$0uLHUH  HUH   HHMHH   rH'HIH+HHv8		 y    AIcLULME~UIHE߄IE|
tHE߄IE|
uAHHE~HUIICAHIc|.DDIcH;UwHEIICHUD$H+UE3HMEMODeHE   LeH>	 HHM'H3x L$   I[@IsHIA_A^A\_]@SH0HAHHH0[HD	 Hz ELuE   H~%zH9AuEB|A   L9BuKH~%zH9Au]HJH@ H  HHH+HHHHi H+i  AD;t'"D;uHRHBHuH;H9AuE2ADHHJ@SH0HIDHT$ H@	 9t3@L@HKMuH~%zHtHAHH;HuH~%zI9PH0[HQHuL;@SH IIDLB	 HH   HH [@SH   H  H3HD$pHEHHT$8A0   LD$@AH@ *	 H# Hc Hc IIB<  uHH  HHL$pH3u HĀ   [̅@SH0HyHHu
HIHH@Hy uHHIHH@(DH	 HH0[LD$LL$ SUVWH8IHl$xHHo   Hl$(LHd$  LHHH	 H8_^][HyLtHIH@ H  HHH+HHHHi H+i  A HAv LD$LL$ SUVWH8IHl$xHHHl$(LHd$  LHHHb	 HH8_^][H\$Hl$ VWATAVAWH LqHHEI+HH;   HiN$2IHH;wHHHH+H;wH)HH;HBHHHHBH   r
  HHt
t 3LfMH^HHrKHH HUE<>AD> H   rHKH'H+HCHw
HHs 	 Hz E<>AD> H>HH\$`Hl$hH A_A^A\_^R  H\$UVWAVAWHH`H H3HEHHHUE3DqEu#L:LzLzEGH?	 H    HL}L}L}HLIF8<uHM~  EL?	 A   AHMHELIF8< uHUHM  DNEt0L|?	 IHMHELIF8< uHUHMs  HvD8>tjA   HI?	 HMT  HD8<>uLHHM<  HMH;MsHAHEHEL9uHCEf' A'DE   HMdEMKHHMH3q H$   H`A_A^_^]H\$Ht$ UWAVHl$H   H H3HE7HHHUe    HyuHIHDH@r	 Dw0Hy uH_A
   HIHDH@(?	 DADuLH=	 H&  H{   He WEA   H=	 HM
  HKHH@	 HLIB<  uHHM  DLw=	     HM!HEHǀ<8 uLHUHM  EEMMADufo8	 EE    H{ u	H<	 HKHH B	 He WEHLIB<  uHHM  H{tD5HKH@ H  HHH+HHHHi H+Di  DL<	     HM5HEHǀ<8 uLHUHM  EEMMADufoL	 EE HUH}HCULEHe     HUHr0HHMHH;rH'HIH+HHv_ )o H{   A   H;	 H  HKHrHH
;	 HU}L@HxrH HH  HUHr0HHMHH;rH'HIH+HHv n A   H;	 H{  HHM7H3Pn L$   I[0Is8IA^_]H\$UVWATAUAVAWH LyHL$   HI+MHH;  LqJ,:HHH;wIHHH+L;wJ1HH;HBHHHHBH   r
  HHtm 3HoI,7H_MHIrTHHl MIH^ IVA, H   rHKH'H+HCHw
HHLm "t H MIH
 A, H7HH\$pH A_A^A]A\_^]
  H\$Ht$WH0HIHIHGH+L;w*HJHGHrHHH褁 H3 DD$@LHHt$ HIH\$HHt$PH0_@USVWAVHl$H   H H3HE'IHHHMHUHLHe WEIIB< uHHM  A   H.9	 HM
 EHMH` H@     MHUHMHxrH H
)	 HHSWHEEHM H	 
